---
title: "Code used to generate paper results : CT characteristics"
author: "Julie Devis"
date: "2023-08-28"
output: 
  html_document:
    theme: cosmo
    df_print : paged
    highlight : default
    toc: yes
    number_sections : TRUE
    toc_depth : 4
---

# Loading data and packages


```{r library, message = FALSE}
library(readxl)
library(readr)
library(CTexploreR)
library(tidyverse)
library(SummarizedExperiment)
library(ComplexHeatmap)
library(circlize)
library(SingleCellExperiment)
library(msigdbr)
library(DOSE)
library(org.Hs.eg.db)
library(clusterProfiler)
library(patchwork)
library(Vennerable)
```


```{r load_data, warning=FALSE, message=FALSE}
load("../CTdata/eh_data/all_genes.rda")
load("../CTdata/eh_data/GTEX_data.rda")
load("../CTdata/eh_data/CCLE_data.rda")
load("../CTdata/eh_data/TCGA_TPM.rda")
load("../CTdata/eh_data/normal_tissues_multimapping_data.rda")
load("../CTdata/eh_data/mean_methylation_in_tissues.rda")
load("../CTdata/eh_data/DAC_treated_cells_multimapping.rda")
load("../CTdata/eh_data/DAC_treated_cells.rda")
load("../CTdata/eh_data/testis_sce.rda")
load("../CTdata/eh_data/FGC_sce.rda")
load("../CTdata/eh_data/oocytes_sce.rda")
load("../CTdata/eh_data/CT_genes.rda")

load("~/cluster/DataSets/FGC/scWGBS/processed_data/scWGBS_FGC_SE.rda")  
load("~/cluster/CBIOProjects/202406_CT_genes_paper/tmp_data/CT_promoter_regions_hg19_GR.rda")


all_genes_clean <- all_genes %>% 
  dplyr::rename(gene = external_gene_name) %>% 
  mutate(Regulation = case_when(regulated_by_methylation ~ "Methylation",
                                !regulated_by_methylation ~ 
                                  "Not methylation")) %>%
  mutate(Chr = case_when(X_linked ~ "X-linked",
                         !X_linked ~ "not X")) %>% 
  mutate(chr_met = case_when(regulated_by_methylation & X_linked ~ "X_met",
                             !regulated_by_methylation & X_linked ~ "X_not_met",
                             regulated_by_methylation & !X_linked ~ "not_X_met",
                             !regulated_by_methylation & !X_linked ~ 
                               "not_X_not_met")) %>% 
  mutate(GTEX_specificity = case_when(GTEX_category == "testis_specific" ~ 
                                        "Testis specific",
                                        GTEX_category == "testis_preferential" ~ 
                                        "Testis preferential",
                                        GTEX_category == "lowly_expressed" ~ 
                                        "Undetectable in GTEX")) %>% 
  mutate(GTEX_specificity = factor(GTEX_specificity, 
                                     levels = c("Testis specific",  
                                                "Undetectable in GTEX", 
                                                "Testis preferential"))) %>% 
  mutate(testis_specificity = case_when(testis_specificity == 
                                          "testis_specific" ~ "Testis specific",
                                        testis_specificity == 
                                          "testis_preferential" ~ 
                                          "Testis preferential")) %>%
  mutate(testis_specificity = factor(testis_specificity, 
                                     levels = c("Testis specific", 
                                                "Testis preferential"))) %>% 
  mutate(chr_met = factor(chr_met, 
                          levels = c("X_met", "not_X_met", "X_not_met", 
                                     "not_X_not_met"))) %>% 
  mutate(met_3_groups = case_when(regulated_by_methylation & X_linked ~ 
                                    "Methylation X-linked",
                                   regulated_by_methylation & !X_linked ~ 
                                    "Methylation not X-linked",
                             !regulated_by_methylation ~ "Not Methylation")) %>% 
  mutate(met_3_groups = factor(met_3_groups, 
                          levels = c("Methylation X-linked", 
                                     "Methylation not X-linked", 
                                     "Not Methylation"))) 

CT_genes <- all_genes %>%
  filter(CT_gene_type == "CT_gene")
CT_genes_types <- all_genes_clean %>% 
  filter(CT_gene_type == "CT_gene")

CTP_genes <- all_genes %>%
  filter(CT_gene_type == "CTP_gene")
CTP_genes_types <- all_genes_clean %>% 
  filter(CT_gene_type == "CTP_gene")
```


Common figures parameters

```{r fig_param}
legends_param <- list(
  labels_gp = gpar(col = "black", fontsize = 4),
  title_gp = gpar(col = "black", fontsize = 5, fontface = "bold"),
  simple_anno_size = unit(0.2, "cm"),
  row_names_gp = gpar(fontsize = 3),
  annotation_name_side = "left",
  border = FALSE,
  border_gp = gpar(lwd = 0.2),
  grid_width = unit(0.2, "cm"),
  grid_height = unit(0.05, "cm"),
  legend_height = unit(1, "cm"),
  use_raster = TRUE,
               raster_device = "CairoPNG",
               raster_quality = 10)



legend_colors <- c("#5E4FA2", "#3288BD", "#66C2A5", "#ABDDA4", "#E6F598",
                   "#FFFFBF", "#FEE08B", "#FDAE61", "#F46D43", "#D53E4F",
                   "#9E0142")

TCGA_colors <- c(
    "BRCA" = "midnightblue", "COAD" = "darkorchid2",
    "ESCA" = "gold", "HNSC" = "deeppink2",
    "LUAD" = "seagreen", "LUSC" = "seagreen3",
    "SKCM" = "red3")

CCLE_colors <- c(
    "Lung" = "seagreen3", "Skin" = "red3",
    "Bile_Duct" = "mediumpurple1", "Bladder" = "mistyrose2",
    "Colorectal" = "plum", "Lymphoma" = "steelblue1",
    "Uterine" = "darkorange4", "Myeloma" = "turquoise3",
    "Kidney" = "thistle4",
    "Pancreatic" = "darkmagenta", "Brain" = "palegreen2",
    "Gastric" = "wheat3", "Breast" = "midnightblue",
    "Bone" = "sienna1", "Head_and_Neck" = "deeppink2",
    "Ovarian" = "tan3", "Sarcoma" = "lightcoral",
    "Leukemia" = "steelblue4", "Esophageal" = "khaki",
    "Neuroblastoma" = "olivedrab1")

DAC_colors <- c(
    "B2-1" = "olivedrab2", "HCT116" = "lightcoral",
    "HEK293T" = "seagreen3", "HMLER" = "mediumpurple1",
    "IMR5-75" = "deeppink2", "NCH1681" = "steelblue2",
    "NCH612" = "red3", "TS603" = "darkmagenta")

DAC_treatment_colors <- c("CTL" = "royalblue4", "5-aza" = "maroon3")

Regulation_colors <- c("Methylation" = "indianred2", 
                       "Not methylation"= "cyan4")

chr_met_colors = c("X_met" = "mediumorchid2",
                   "X_not_met" = "mediumseagreen",
                   "not_X_met"= "darkorange1",
                   "not_X_not_met" = "dodgerblue1")

Chr_colors = c("X-linked" = "deeppink3", "not X" = "lightgreen")
Chr_colors_with_Y = c("X-linked" = "deeppink3", "Autosome" = "lightgreen", 
                      "Y-linked" = "black")

GTEX_specificity_colors = c("Testis specific" = "mediumpurple4",
                            "Testis preferential" = "goldenrod3",
                            "Undetectable in GTEX" = "gray")

Testis_specificity_colors = c("Testis specific" = "mediumpurple2",
                              "Testis preferential" = "goldenrod1",
                              "Undetectable in GTEX" = "gray")

testis_colors <- c(
    "SSC" = "floralwhite", "Spermatogonia" = "moccasin",
    "Early_spermatocyte" = "gold",
    "Late_spermatocyte" = "orange",
    "Round_spermatid" = "red2",
    "Elongated_spermatid" = "darkred", "Sperm" = "purple",
    "Sertoli" = "gray",
    "Leydig" = "cadetblue2", "Myoid" = "springgreen3",
    "Macrophage" = "gray10",
    "Endothelial" = "steelblue")

Fetal_cell_colors = c( "F_PGC" = "pink", 
                       "F_GC" = "pink3", 
                       "F_oogonia" = "palevioletred3",
                       "F_oocyte" = "mediumorchid4", 
                       "M_PGC" = "lightblue1", 
                       "M_GC" = "steelblue3",
                       "M_pre_spermatogonia" = "royalblue3")

Fetal_cellType_colors <- c( 
"F_GC" = "pink3",
  "F_oogonia" = "palevioletred3",
  "F_oocyte" = "mediumorchid4",
  "M_GC" = "steelblue3",
  "M_pre_spermatogonia" = "royalblue3",
  "F_soma" = "gray",
  "M_soma" = "gray")

Fetal_time_colors <- c("6W" = "aliceblue",
                      "7W" = "cyan",
                      "8W" = "cyan4",
                      "9W" = "skyblue1",
                      "10W" = "dodgerblue",
                      "17W" = "blue",
                      "21W" = "lightblue4",
                      "24W" = "black")

oocytes_colors <- c("Growing oocytes" = "mediumpurple1",
                    # "Fully grown oocytes" ="lightgreen",
                     # "Metaphase I" = "mediumvioletred",
                    "Metaphase II" = "mediumpurple4")

sex_colors <- c("F" = "palevioletred1", "M"= "skyblue1")

stage_colors <- c( "pre-meiotic" = "peachpuff", 
                   "meiotic"= "lightsalmon",
                   "post-meiotic" = "maroon4",
                   "somatic" = "gray")

```

# CT selection

## Pipeline figure values

Here is the code used to have all the values in the pipeline figure.

As a reminder testis_specificity has been done summarizing testis-specificity 
analysis from GTEX, multimapping, HPA_cell_type_specificities, CCLE and TCGA 
categories.

Testis-specific genes must be: classified as "testis-specific" in GTEX or 
multimappig analysis +  must not be detected in any somatic cell type (in HPA 
scRNAseq analysis) + can not be classified as "leaky" in CLLE and TCGA 
categories + can not be detected (TPM > 0.5) in more than 75% of normal 
peritumoral tissues from a TCGA tumor type.

Testis_preferential genes must be: classified as "testis_preferential" in GTEX 
or multimappig analysis or initially classified as "testis-specific" in GTEX or 
multimappig analysis but downgraded to "testis_preferential because they were 
detected in a somatic cell type (in HPA scRNAseq analysis) + can be detected in 
some somatic cell type but the level has to be at least 10 times lower than the
level detected in a germ cell type (in HPA scRNAseq analysis) must not be 
classified as "leaky" in CLLE and TCGA categories.

```{r pipeline figures value}
table(all_genes$GTEX_category)
table(all_genes$multimapping_analysis)

table(all_genes$testis_specificity)
table(all_genes$GTEX_category, all_genes$testis_specificity)

specific_to_pref <- all_genes %>%
  filter(GTEX_category == "testis_specific" |
           multimapping_analysis == "testis_specific") %>% 
  filter(testis_specificity == "testis_preferential")
# No more specific to pref as their ratio are all < 10
# all_genes %>%
#   filter(GTEX_category == "testis_specific" |
#            multimapping_analysis == "testis_specific") %>% 
#   filter(!not_detected_in_somatic_HPA)

# Genes that were selected with GTEx but leaky so removed (ubiquitary expression)
all_genes %>% 
  filter(GTEX_category %in% c("testis_specific", "testis_preferential")) %>% 
  filter(TCGA_category == "leaky" | CCLE_category == "leaky") %>% 
  dim()

specific_to_none <- all_genes %>%
  filter(GTEX_category == "testis_specific" |
           multimapping_analysis == "testis_specific") %>% 
  filter(testis_specificity == "not_testis_specific")
table(specific_to_none$CCLE_category)
table(specific_to_none$TCGA_category)
table(specific_to_none$max_q75_in_NT < 0.5)
# 95 lost at this step
table(specific_to_none$not_detected_in_somatic_HPA)
table(specific_to_none$HPA_ratio_germ_som >= 10)

preferential_to_none <- all_genes %>%
  filter(GTEX_category == "testis_preferential" |
           multimapping_analysis == "testis_preferential") %>% 
  filter(testis_specificity == "not_testis_specific")
table(preferential_to_none$CCLE_category)
table(preferential_to_none$TCGA_category)
table(preferential_to_none$not_detected_in_somatic_HPA)
# Almost all detected in somatic
table(preferential_to_none$HPA_ratio_germ_som >= 10)


all_genes %>% 
  filter(testis_specificity != "not_testis_specific") %>% 
  filter(CCLE_category == "activated" & 
           (TCGA_category == "multimapping_issue" | 
              TCGA_category == "activated")) %>% 
  pull(testis_specificity) %>% 
  table()

table(all_genes$IGV_backbone)
table(all_genes$CT_gene_type)
```

## GTEX-TCGA-CCLE

Illustrate the selection using heatmaps with GTEx data, TCGA and CCLE

Function to print GTEX expression with specific annotations.

```{r GTEX_function}
print_GTEX_heatmap <- function(genes = NULL, units = "log_TPM", 
                               row_fontsize = 2,
                               column_fontsize = 6, 
                               h_width = 3,
                               h_height = 10,
                               split_rows_by = "regulation",
                               clust_method = "ward.D",
                               clustRow = FALSE) {
  
  suppressMessages({
    database <- GTEX_data
  })
  
  database <- database[rowData(database)$external_gene_name %in% genes, ]
  
  mat <- assay(database)
  rownames(mat) <- rowData(database)$external_gene_name
  
  if (units == "log_TPM") mat <- log1p(mat)
    
  gene_type <- all_genes_clean %>% 
    filter(gene %in% genes) %>% 
    mutate(gene = factor(gene, levels = genes)) %>% 
    arrange(gene)
   

    if (split_rows_by == "regulation") {
        if (clustRow) gene_type <- gene_type %>% 
            arrange(Regulation)
      rowSplit <- gene_type$Regulation
    }
    if (split_rows_by == "testis_specificity") {
        if (clustRow) gene_type <- gene_type %>% 
            arrange(testis_specificity)
      rowSplit <- gene_type$testis_specificity
    }
    if (split_rows_by == "chr_met") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(chr_met)
      rowSplit <- gene_type$chr_met
    }
    if (split_rows_by == "GTEX_specificity") {
      if (clustRow) GTEX_category <- gene_type %>% 
          arrange(GTEX_specificity)
      rowSplit <- gene_type$GTEX_specificity
    }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(Regulation = Regulation,
                                     annotation_legend_param = legends_param,
                                     simple_anno_size = unit(0.2, "cm"),
                                     col = list(Regulation = Regulation_colors),
                                     annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(chr_met = chr_met,
                                  annotation_legend_param = legends_param,
                                  simple_anno_size = unit(0.2, "cm"),
                                  col = list(chr_met = chr_met_colors),
                                  annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$Chr)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              simple_anno_size = unit(0.2, "cm"),
                              col = list(Chr = Chr_colors),
                              annotation_name_gp = gpar(fontsize = 0))
  
  GTEX_specificity <- as.matrix(gene_type$GTEX_specificity)
  rownames(GTEX_specificity) <- gene_type$gene
  row_ha_GTEX <- rowAnnotation(specificity = GTEX_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(specificity = 
                                          GTEX_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  testis_specificity <- as.matrix(gene_type$testis_specificity)
  rownames(testis_specificity) <- gene_type$gene
  row_ha_TS <- rowAnnotation(testis_specificity = testis_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(testis_specificity = 
                                          Testis_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  heatmap_colors <- colorRamp2(
    seq(0, 5, length = 11),                 
    legend_colors)
  
  h <- Heatmap(mat[as.vector(gene_type$gene), ],
               name = "log(TPM +1)",
               column_title = "Normal tissues (GTEx)",
               column_title_gp = gpar(fontsize = 6, face = "bold"),
               row_split = rowSplit,
               cluster_row_slices = FALSE,
               row_title_gp = gpar(fontsize = 6, face = "bold"),
               border = TRUE,
               border_gp = gpar(lwd = 0.5),
               col = heatmap_colors,
               cluster_rows = clustRow,
               cluster_columns = TRUE,
               show_row_dend = FALSE,
               show_column_dend = FALSE,
               row_names_gp = gpar(fontsize = row_fontsize),
               column_names_gp = gpar(fontsize = column_fontsize),
               column_names_side = "top",
               # right_annotation = c(row_ha_regulation, 
               #                        row_ha_chr, 
               #                        row_ha_ts,
               #                        gap = unit(1, "mm")),
               row_names_side = "left",
               width = unit(h_width, "cm"),
               height = unit(h_height, "cm"),
               clustering_method_rows = clust_method,
               use_raster = TRUE,
               raster_device = "CairoPNG",
               raster_quality = 10,
               heatmap_legend_param = legends_param)
  
  return(h)
}
```

Function to print TCGA expression with specific annotations and to downsample 
the number of tumor samples per tumor type.

```{r TCGA_function}
print_TCGA_heatmap <- function(genes = NULL,
                               tum_type = c("TCGA-SKCM", "TCGA-LUAD", 
                                            "TCGA-LUSC", "TCGA-COAD", 
                                            "TCGA-BRCA", "TCGA-HNSC",
                                            "TCGA-ESCA"),
                               downsample_to_n = 450,
                               fontsize = 4, 
                               clust_method = "ward.D",
                               split_rows_by = "regulation", 
                               h_width = 5, 
                               h_height = 10,
                               show_annotation_legend = TRUE,
                               show_top_annotations = TRUE,
                               show_row_names = FALSE,
                               clustRow = FALSE,
                               clustCol = TRUE) {
  
  tcga <- TCGA_TPM
  tcga <- tcga[,tcga$shortLetterCode != "NT"]
  
  ## downsample tumors to have the same number of tumors in each type
  TCGA <- tcga[, tcga$project_id == tum_type[1]]
  TCGA <- TCGA[, 1:min(ncol(TCGA), downsample_to_n)]
  
  for (tum in tum_type[-1]){
    tmp <- tcga[, tcga$project_id == tum]
    tmp <- tmp[, 1:min(ncol(tmp), downsample_to_n)]
    TCGA <- cbind(TCGA, tmp)
  }
  
  TCGA$project_id <- gsub(pattern = "TCGA-", x = TCGA$project_id, 
                          replacement = '')
  
  split_by <- factor(TCGA$project_id)
  
  column_ha_tumor <- HeatmapAnnotation(
    TCGA_tumor = TCGA$project_id,
    border = TRUE,
    col = list(TCGA_tumor = TCGA_colors),
    annotation_label = gpar(fontsize = 4),
    annotation_name_gp = gpar(fontsize = 0),
    simple_anno_size = unit(0.3, "cm"),
    annotation_legend_param = legends_param)
  
  ## Use gene names instead of ENSEMBL IDs
  mat <- SummarizedExperiment::assay(TCGA)
  rownames(mat) <- rowData(TCGA)$external_gene_name
  mat <- log1p(mat)
  
  gene_type <- all_genes_clean %>% 
    filter(gene %in% genes) %>% 
    mutate(gene = factor(gene, levels = genes)) %>% 
    arrange(gene)
  
  
  rowSplit <- NULL
  if(!is.null(split_rows_by)){
    if (split_rows_by == "regulation") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(Regulation)
      rowSplit <- gene_type$Regulation
    }
    if (split_rows_by == "testis_specificity") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(testis_specificity)
      rowSplit <- gene_type$testis_specificity
    }
    if (split_rows_by == "GTEX_specificity") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(GTEX_specificity)
      rowSplit <- gene_type$GTEX_specificity
    }
    
    if (split_rows_by == "chr_met") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(chr_met)
      rowSplit <- gene_type$chr_met
    }
  }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(Regulation = Regulation,
                                     annotation_legend_param = legends_param,
                                     simple_anno_size = unit(0.2, "cm"),
                                     col = list(Regulation = Regulation_colors),
                                     annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(chr_met = chr_met,
                                  annotation_legend_param = legends_param,
                                  simple_anno_size = unit(0.2, "cm"),
                                  col = list(chr_met = chr_met_colors),
                                  annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$Chr)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              simple_anno_size = unit(0.2, "cm"),
                              col = list(Chr = Chr_colors),
                              annotation_name_gp = gpar(fontsize = 0))
  
  GTEX_specificity <- as.matrix(gene_type$GTEX_specificity)
  rownames(GTEX_specificity) <- gene_type$gene
  row_ha_GTEX <- rowAnnotation(GTEX_specificity = GTEX_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(GTEX_specificity = 
                                          GTEX_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  testis_specificity <- as.matrix(gene_type$testis_specificity)
  rownames(testis_specificity) <- gene_type$gene
  row_ha_TS <- rowAnnotation(specificity = testis_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(testis_specificity = 
                                          Testis_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  rAnnot <- NULL
  if(show_annotation_legend){
    if (split_rows_by == "regulation") {
  rAnnot <- c(row_ha_regulation, gap = unit(1, "mm"))
    }
    if (split_rows_by == "testis_specificity") {
  rAnnot <- c(row_ha_TS, gap = unit(1, "mm"))
    }
    if (split_rows_by == "GTEX_specificity") {
  rAnnot <- c(row_ha_GTEX, gap = unit(1, "mm"))
    }
    
    if (split_rows_by == "chr_met") {
  rAnnot <- c(row_ha_TS, gap = unit(1, "mm"))
    }
  }

  
  # same range as in GTEX
  heatmap_colors <- colorRamp2(
    seq(0, 5, length = 11),                 
    legend_colors)
  
  tAnnot <- NULL
  if (show_top_annotations) tAnnot <- column_ha_tumor
  
  Heatmap(mat[as.vector(gene_type$gene), ],
          name = "TCGA (logTPM)",
          column_title = "TCGA tumor samples",
          column_title_gp = gpar(fontsize = 6, face = "bold"),
          clustering_method_rows = clust_method,
          clustering_method_columns = clust_method,
          column_split = split_by,
          row_gap = unit(c(1), "mm"),
          col = heatmap_colors,
          cluster_rows = clustRow,
          cluster_columns = clustCol,
          show_row_names = show_row_names,
          show_column_names = FALSE,
          show_row_dend = FALSE,
          show_column_dend = FALSE,
          column_names_gp = gpar(fontsize = 6),
          column_names_centered = TRUE,
          row_names_gp = gpar(fontsize = fontsize),
          row_names_side = "left",
          border = TRUE,
          border_gp = gpar(lwd = 0.5),
          column_names_side = c("top"),
          column_names_rot = 90,
          row_title_gp = gpar(fontsize = 0),
          top_annotation = tAnnot,
          right_annotation = rAnnot,
          row_split = rowSplit,
          cluster_row_slices = FALSE,
          width = unit(h_width, "cm"),
          height = unit(h_height, "cm"),
          use_raster = TRUE,
          raster_device = "CairoPNG",
          raster_quality = 10,
          heatmap_legend_param = legends_param,
          show_heatmap_legend = FALSE)
}
```

Function to print CCLE expression with specific annotations. Allows to 
downsample the number of tumor samples pet tumor type to show

```{r CCLE_function}
print_CCLE_heatmap <- function(genes = NULL,
                               tum_type = c("Skin", "Lung", 
                                            "Colorectal", 
                                            "Breast", 
                                            "Head_and_Neck",
                                            "Esophageal"),
                               h_width = 5,
                               h_height = 10,
                               downsample_to_n = 450,
                               fontsize = 4, 
                               clust_method = "ward.D",
                               split_rows_by = "regulation", 
                               show_annotation_legend = TRUE,
                               clustRow = FALSE,
                               clustCol = TRUE){
  
  ## downsample tumors to have the same number of tumors in each type
  ccle <- CCLE_data
  CCLE <- ccle[, ccle$type == tum_type[1]]
  CCLE <- CCLE[, 1:min(ncol(CCLE), downsample_to_n)]
  
  for (tum in tum_type[-1]){
    tmp <- ccle[, ccle$type == tum]
    tmp <- tmp[, 1:min(ncol(tmp), downsample_to_n)]
    CCLE <- cbind(CCLE, tmp)
  }
  
  split_by <- factor(CCLE$type)
  
  column_ha_tumor <- HeatmapAnnotation(
    CCLE_cell_type = CCLE$type,
    border = TRUE,
    col = list(CCLE_cell_type = CCLE_colors),
    annotation_label = gpar(fontsize = 4),
    annotation_name_gp = gpar(fontsize = 0),
    simple_anno_size = unit(0.3, "cm"),
    annotation_legend_param = legends_param)
  
  ## Use gene names instead of ENSEMBL IDs
  mat <- SummarizedExperiment::assay(CCLE)
  rownames(mat) <- rowData(CCLE)$external_gene_name
  mat <- log1p(mat)
  
    
  gene_type <- all_genes_clean %>% 
    filter(gene %in% genes) %>% 
    mutate(gene = factor(gene, levels = genes)) %>% 
    arrange(gene)
   
 rowSplit <- NULL
  if(!is.null(split_rows_by)){
    if (split_rows_by == "regulation") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(Regulation)
      rowSplit <- gene_type$Regulation
    }
    if (split_rows_by == "testis_specificity") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(testis_specificity)
      rowSplit <- gene_type$testis_specificity
    }
    if (split_rows_by == "GTEX_specificity") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(GTEX_specificity)
      rowSplit <- gene_type$GTEX_specificity
    }
    
    if (split_rows_by == "chr_met") {
      if (clustRow) gene_type <- gene_type %>% 
          arrange(chr_met)
      rowSplit <- gene_type$chr_met
    }
  }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(Regulation = Regulation,
                                     annotation_legend_param = legends_param,
                                     simple_anno_size = unit(0.2, "cm"),
                                     col = list(Regulation = Regulation_colors),
                                     annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(chr_met = chr_met,
                                  annotation_legend_param = legends_param,
                                  simple_anno_size = unit(0.2, "cm"),
                                  col = list(chr_met = chr_met_colors),
                                  annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$Chr)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              simple_anno_size = unit(0.2, "cm"),
                              col = list(Chr = Chr_colors),
                              annotation_name_gp = gpar(fontsize = 0))
  
  GTEX_specificity <- as.matrix(gene_type$GTEX_specificity)
  rownames(GTEX_specificity) <- gene_type$gene
  row_ha_GTEX <- rowAnnotation(GTEX_specificity = GTEX_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(GTEX_specificity = 
                                          GTEX_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  testis_specificity <- as.matrix(gene_type$testis_specificity)
  rownames(testis_specificity) <- gene_type$gene
  row_ha_TS <- rowAnnotation(specificity = testis_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(testis_specificity = 
                                          Testis_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  rAnnot <- NULL
  if(show_annotation_legend){
    if (split_rows_by == "regulation") {
  rAnnot <- c(row_ha_regulation, gap = unit(1, "mm"))
    }
    if (split_rows_by == "testis_specificity") {
  rAnnot <- c(row_ha_TS, gap = unit(1, "mm"))
    }
    if (split_rows_by == "GTEX_specificity") {
  rAnnot <- c(row_ha_GTEX, gap = unit(1, "mm"))
    }
    
    if (split_rows_by == "chr_met") {
  rAnnot <- c(row_ha_TS, gap = unit(1, "mm"))
    }
  }

  
  # Same range as GTEX and TCGA
  heatmap_colors <- colorRamp2(
    seq(0, 5, length = 11),                 
    legend_colors)
  
  Heatmap(mat[as.vector(gene_type$gene), ],
          name = "CCLE (logTPM)",
          column_title = paste0("CCLE tumor cell lines"),
          column_title_gp = gpar(fontsize = 6, face = "bold"),
          clustering_method_rows = clust_method,
          clustering_method_columns = clust_method,
          column_split = split_by,
          row_gap = unit(c(1), "mm"),
          col = heatmap_colors,
          cluster_rows = clustRow,
          cluster_columns = clustCol,
          show_row_names = TRUE,
          show_column_names = FALSE,
          show_row_dend = FALSE,
          show_column_dend = FALSE,
          column_names_gp = gpar(fontsize = 6),
          column_names_centered = TRUE,
          row_names_gp = gpar(fontsize = fontsize),
          row_names_side = "left",
          border = TRUE,
          border_gp = gpar(lwd = 0.5),
          column_names_side = c("top"),
          column_names_rot = 90,
          row_title_gp = gpar(fontsize = 0),
          top_annotation = column_ha_tumor,
          right_annotation = rAnnot,
          row_split = rowSplit,
          width = unit(h_width, "cm"),
          height = unit(h_width, "cm"),
          use_raster = TRUE,
          raster_device = "CairoPNG",
          raster_quality = 10,
          heatmap_legend_param = legends_param,
          show_heatmap_legend = FALSE)
}
```

Show CT genes in GTEX, CCLE and TCGA (downsampling the number of samples by 
tumor type).

```{r selection_heatmap}
my_genes <- all_genes %>% 
  filter(CT_gene_type != "other") %>% 
  pull(external_gene_name)

# Run the heatmap with custRow parameters to define the correct order
# This order will then be used for GTEX and CCLE heatmaps
ht_TCGA <- print_TCGA_heatmap(my_genes, 
                              fontsize = 4, 
                              downsample_to_n = 50, 
                              h_width = 3,
                              h_height = 8,
                              tum_type = c("TCGA-SKCM", "TCGA-LUAD", 
                                           #"TCGA-ESCA", "TCGA-LUSC", 
                                           #"TCGA-COAD",
                                           #"TCGA-BRCA",
                                           "TCGA-HNSC"),
                              clust_method = "ward.D",
                              clustRow = TRUE,
                              split_rows_by = "testis_specificity")
indices <- row_order(ht_TCGA)
gene_order <- all_genes_clean %>%
  filter(gene %in% my_genes) %>%
  arrange(testis_specificity) %>%
  pull(gene)

TS_order <- gene_order[indices$`Testis specific`]
TP_order <- gene_order[indices$`Testis preferential`]
my_genes <- c(TS_order, TP_order)

ht_TCGA <- print_TCGA_heatmap(my_genes, 
                              fontsize = 4, 
                              downsample_to_n = 50, 
                              h_width = 3,
                              h_height = 8,
                              tum_type = c("TCGA-SKCM", "TCGA-LUAD", 
                                           #"TCGA-ESCA", "TCGA-LUSC", 
                                           #"TCGA-COAD",
                                           #"TCGA-BRCA",
                                           "TCGA-HNSC"),
                              clust_method = "ward.D",
                              show_annotation_legend = FALSE,
                              clustRow = FALSE,
                              split_rows_by = "testis_specificity")

ht_CCLE <- print_CCLE_heatmap(my_genes, 
                              downsample_to_n = 50, 
                              tum_type = c("Skin", "Lung", 
                                           #"colorectal", 
                                           #"breast", 
                                           #"esophageal",
                                           "Head_and_Neck"),
                              #"Leukemia", "Lymphoma"),
                              clustRow = FALSE,
                              show_annotation_legend = FALSE,
                              h_width = 3,
                              h_height = 8,
                              split_rows_by = "testis_specificity")


ht_GTEX <- print_GTEX_heatmap(my_genes, 
                              row_fontsize = 1,
                              column_fontsize = 5,
                              h_width = 5,
                              h_height = 8,
                              clustRow = FALSE,
                              split_rows_by = "testis_specificity")

h_list <- ht_GTEX + ht_CCLE + ht_TCGA
draw(h_list, main_heatmap = "log(TPM +1)", ht_gap = unit(c(0.3, 0.01), "cm"),
     merge_legend = TRUE)
```

```{r, eval = FALSE, echo=FALSE}
pdf(file = paste0("figs/GTEX_Tumors.pdf"))
draw(h_list, main_heatmap = "log(TPM +1)", 
     ht_gap = unit(c(0.3, 0.01), "cm"),
     merge_legend = TRUE)
dev.off()
```


## Multimapping Fig

Figure showing how we kept the lowly_expressed in GTEx

```{r multimapping}
my_genes <- all_genes_clean %>% 
  filter(GTEX_category == 'lowly_expressed')  %>% 
  filter(CT_gene_type != 'other')  %>% 
    arrange(testis_specificity)
rowSplit <- my_genes$testis_specificity

testis_specificity <- as.matrix(my_genes$testis_specificity)
rownames(testis_specificity) <- my_genes$gene
row_ha_TS <- rowAnnotation(testis_specificity = testis_specificity,
                           annotation_legend_param = legends_param,
                           simple_anno_size = unit(0.2, "cm"),
                           col = list(testis_specificity = 
                                        Testis_specificity_colors),
                           annotation_name_gp = gpar(fontsize = 0))

without_MP <- log1p(
  assay(normal_tissues_multimapping_data[my_genes$ensembl_gene_id,], 
    "TPM_no_multimapping"))
with_MP <-  log1p(
  assay(normal_tissues_multimapping_data[my_genes$ensembl_gene_id,], 
    "TPM_with_multimapping"))
rownames(without_MP) <- rownames(with_MP) <- 
  rowData(normal_tissues_multimapping_data)[my_genes$ensembl_gene_id, 
                                            ]$external_gene_name
# Fix same range of colors for both heatmaps
full_mat <- cbind(without_MP, with_MP)
heatmap_colors <- colorRamp2(
  seq(0, quantile(rowMax(full_mat), 0.8), length = 11), legend_colors)


colnames(without_MP) <- gsub("_", x = str_to_sentence(colnames(without_MP)),
                             ' ')
colnames(with_MP) <- gsub("_", x = str_to_sentence(colnames(with_MP)), ' ')
ordered_tissues <- 
  c("Testis", colnames(without_MP)[- which(colnames(without_MP) == "Testis")])
h_without_MP <- Heatmap(without_MP[my_genes$gene, ordered_tissues],
                        name = "no MP",
                        column_title = "Multimapped reads excluded",
                        column_title_gp = gpar(fontsize = 10, face = "bold"),
                        row_title_gp = gpar(fontsize = 0, face = "bold", 
                                            angle  = 90),
                        row_split = rowSplit,
                        border = TRUE,
                        border_gp = gpar(lwd = 0.5),
                        col = heatmap_colors,
                        cluster_rows = FALSE,
                        cluster_columns = FALSE,
                        show_row_dend = FALSE,
                        show_column_dend = FALSE,
                        row_names_gp = gpar(fontsize = 4),
                        column_names_gp = gpar(fontsize = 6),
                        column_names_side = "top",
                        row_names_side = "left",
                        heatmap_width = unit(8, "cm"),
                        heatmap_height = unit(12, "cm"),
                        clustering_method_rows = "ward.D",
                        use_raster = TRUE,
                        raster_device = "CairoPNG",
                        raster_quality = 10,
                        show_heatmap_legend = FALSE,
                        heatmap_legend_param = legends_param)

h_with_MP <- Heatmap(with_MP[my_genes$gene, ordered_tissues],
                     name = "log(TPM)",
                     column_title = "Multimapped reads counted",
                     column_title_gp = gpar(fontsize = 10, face = "bold"),
                     row_title_gp = gpar(fontsize = 0, face = "bold"),
                     row_split = rowSplit,
                     border = TRUE,
                     border_gp = gpar(lwd = 0.5),
                     col = heatmap_colors,
                     cluster_rows = TRUE,
                     cluster_columns = FALSE,
                     show_row_dend = FALSE,
                     show_column_dend = FALSE,
                     row_names_gp = gpar(fontsize = 4),
                     column_names_gp = gpar(fontsize = 6),
                     column_names_side = "top",
                     row_names_side = "left",
                     heatmap_width = unit(8, "cm"),
                     heatmap_height = unit(12, "cm"),
                     clustering_method_rows = "ward.D",
                     right_annotation = c(row_ha_TS, gap = unit(1, "mm")),
                     use_raster = TRUE,
                     raster_device = "CairoPNG",
                     raster_quality = 10,
                     heatmap_legend_param = legends_param)

h_list <- h_without_MP + h_with_MP
draw(h_list, main_heatmap = "no MP", ht_gap = unit(0.3, "cm"),
                 height = unit(5, "cm"))
```

```{r, eval = FALSE, echo = FALSE}
pdf(file = paste0("figs/Multimapping.pdf"))
draw(h_list, main_heatmap = "no MP", ht_gap = unit(0.3, "cm"),
                 height = unit(5, "cm"))
dev.off()
```

## IGV

Shows why we removed illegitimate transcripts

```{r, echo=FALSE, fig.align='left', out.width = '100%', purl = TRUE}
knitr::include_graphics("figs/Fig_IGV_no_clear_transcript.png")
```

# Methylation analysis

## Regulation by methylation : WGBS + DAC heatmap

```{r meth_heatmap_function}
met_genes <- CT_genes %>% 
  filter(regulated_by_methylation) %>% 
  pull(external_gene_name)

met_value_number_per_gene <- tibble(
  gene = CT_genes$external_gene_name,
  regulated_by_methylation = CT_genes$regulated_by_methylation,
  NA_met_values_number = rowSums(
    is.na(assay(mean_methylation_in_tissues[CT_genes$external_gene_name, ]))))
n_met_values <- ncol(mean_methylation_in_tissues)

CT_met_with_WGBS <- met_value_number_per_gene %>%
  filter(regulated_by_methylation) %>% 
  filter(NA_met_values_number !=  n_met_values)

CT_met_without_WGBS <- met_value_number_per_gene %>%
  filter(regulated_by_methylation) %>% 
  filter(NA_met_values_number == n_met_values)

CT_not_met_with_WGBS <- met_value_number_per_gene %>%
  filter(!regulated_by_methylation) %>% 
  filter(NA_met_values_number !=  n_met_values)

CT_not_met_without_WGBS <- met_value_number_per_gene %>%
  filter(!regulated_by_methylation) %>% 
  filter(NA_met_values_number == n_met_values)

mean_methylation_fig <- function(genes = NULL, 
                                 values_only = FALSE,
                                 row_fontsize = 4,
                                 column_fontsize = 8, 
                                 na.omit = FALSE) {
  
  database <- mean_methylation_in_tissues
  rowData(database)$external_gene_name <- rownames(database)
  database <- database[genes, ]
  
  if (na.omit) {
    mat <- na.omit(assay(database))
    clustering_option <- TRUE
  } else {
    mat <- assay(database)
    clustering_option <- FALSE
  }
  
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) %>% 
    mutate(group = case_when(gene %in% CT_met_with_WGBS$gene ~ "group1",
                             gene %in% CT_met_without_WGBS$gene ~ "group2",
                             gene %in% CT_not_met_with_WGBS$gene ~ "group3",
                             gene %in% CT_not_met_without_WGBS$gene ~ 
                               "group4")) %>% 
    left_join(CT_genes %>% dplyr::rename(gene = external_gene_name) %>% 
                dplyr::select(gene, somatic_met_level, sperm_met_level, 
                              germline_methylation, somatic_methylation)) %>% 
    arrange(group, sperm_met_level)

  # sort genes regulated by met differently than not regulated by met
  gene_type_met <- gene_type %>%
    filter(regulated_by_methylation) 
    
  gene_type_not_met <- gene_type %>%
    filter(!regulated_by_methylation) %>%
    arrange(germline_methylation, somatic_met_level)

  gene_type <- rbind(gene_type_met, gene_type_not_met)
  colnames(mat) <- str_to_sentence(colnames(mat))
  
  h <- Heatmap(mat[gene_type$gene, sort(colnames(mat))],
               column_title = "Promoter methylation",
               name = "Mean \nmethylation",
               column_title_gp = gpar(fontsize = 6),
               border = TRUE,
               border_gp = gpar(lwd = 0.5),
               row_title_gp = gpar(col = Regulation_colors, fontsize = 6),
               row_split = factor(gene_type$Regulation,
                                  levels = c("Methylation", "Not methylation")),
               #row_split = factor(gene_type$group),
               cluster_row_slices = FALSE,
               col = colorRamp2(seq_len(100),
                                colorRampPalette(c("moccasin", 
                                                   "dodgerblue4"))(100)),
               na_col = "gray80",
               cluster_rows = FALSE,
               cluster_columns = FALSE,
               show_row_names = TRUE,
               show_heatmap_legend = TRUE,
               show_row_dend = FALSE,
               column_names_gp = gpar(fontsize = column_fontsize),
               row_names_gp = gpar(fontsize = row_fontsize),
               row_names_side = "left",
               heatmap_width = unit(4, "cm"),
               heatmap_height = unit(10, "cm"),
               use_raster = TRUE,
               raster_device = "CairoPNG",
               raster_quality = 8,
               column_names_side = "top",
               heatmap_legend_param = legends_param)
  
  return(h)
}
```

```{r DAC_heatmap_function}
DAC_fig <- function(genes = gene_type$gene, 
                    multimapping = TRUE, 
                    row_fontsize = 2,
                    units = "log1p",
                    values_only = FALSE, scale_lims = NULL) {
  

  if (multimapping) {
    # database <- CTdata::DAC_treated_cells_multimapping()
    database <- DAC_treated_cells_multimapping
    
  } else {
    #database <- CTdata::DAC_treated_cells()
    database <- DAC_treated_cells
  }
  
  database$treatment[database$treatment == "DAC"] <- "5-aza"
  database$treatment <- factor(database$treatment, levels = c("CTL", "5-aza"))
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) %>% 
    mutate(group = case_when(gene %in% CT_met_with_WGBS$gene ~ "group1",
                             gene %in% CT_met_without_WGBS$gene ~ "group2",
                             gene %in% CT_not_met_with_WGBS$gene ~ "group3",
                             gene %in% CT_not_met_without_WGBS$gene ~ "group4")) %>% 
    left_join(CT_genes %>% dplyr::rename(gene = external_gene_name) %>% 
                dplyr::select(gene, somatic_met_level, sperm_met_level, 
                              germline_methylation, somatic_methylation)) %>% 
    arrange(group, sperm_met_level)

  # sort genes regulated by met differently than not regulated by met
  gene_type_met <- gene_type %>%
    filter(regulated_by_methylation) 
    
  gene_type_not_met <- gene_type %>%
    filter(!regulated_by_methylation) %>%
    arrange(germline_methylation, somatic_met_level)

  gene_type <- rbind(gene_type_met, gene_type_not_met)
  rownames(database) <- rowData(database)$external_gene_name
  
  mat <- assay(database, units)
  
  df_col <- data.frame("cell" = colData(database)$cell,
                       "treatment" = colData(database)$treatment)
  rownames(df_col) <- rownames(colData(database))
  df_col <- df_col[order(df_col$cell, df_col$treatment), ]
  
  column_ha_cell <- HeatmapAnnotation(
    cell = df_col$cell,
    border = FALSE, 
    #border_gp = gpar(lwd = 0.1),
    simple_anno_size = unit(0.2, "cm"),
    annotation_legend_param = legends_param,
    col = list("cell" = DAC_colors),
    annotation_name_gp = gpar(fontsize = 0))  
  
  column_ha_treatment <- HeatmapAnnotation(
    treatment = df_col$treatment,
    col = list(treatment = DAC_treatment_colors),
    annotation_legend_param = legends_param,
    border = FALSE, 
    simple_anno_size = unit(0.2, "cm"),
    annotation_name_gp = gpar(fontsize = 0))  
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(Regulation = Regulation,
                                     annotation_legend_param = legends_param,
                                     simple_anno_size = unit(0.2, "cm"),
                                     col = list(Regulation = Regulation_colors),
                                     annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(chr_met = chr_met,
                                  annotation_legend_param = legends_param,
                                  simple_anno_size = unit(0.2, "cm"),
                                  col = list(chr_met = chr_met_colors),
                                  annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$Chr)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              simple_anno_size = unit(0.2, "cm"),
                              col = list(Chr = Chr_colors),
                              annotation_name_gp = gpar(fontsize = 0))
  
  testis_specificity <- as.matrix(gene_type$testis_specificity)
  rownames(testis_specificity) <- gene_type$gene
  row_ha_TS <- rowAnnotation(testis_specificity = testis_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(testis_specificity = 
                                          Testis_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  if (is.null(scale_lims)) scale_lims <- c(0, quantile(rowMaxs(mat), 0.75))
  
  h <- Heatmap(mat[gene_type$gene, rownames(df_col), drop = FALSE],
               name = "logCounts",
               row_split = factor(gene_type$Regulation, 
                                  levels = c("Methylation", "Not methylation")),
               row_title_gp = gpar(col = Regulation_colors, fontsize = 6),
               border = TRUE,
               border_gp = gpar(lwd = 0.5),
               column_title = "DAC induction in cell lines",
               column_title_gp = gpar(fontsize = 6),
               column_split = factor(df_col$cell),
               col = colorRamp2(seq(scale_lims[1], scale_lims[2], length = 11),
                                legend_colors),
               cluster_rows = TRUE,
               show_row_dend = FALSE,
               show_row_names = FALSE,
               clustering_method_rows = "ward.D",
               show_column_names = FALSE,
               cluster_columns = FALSE,
               row_names_gp = gpar(fontsize = row_fontsize),
               top_annotation = c(column_ha_cell, column_ha_treatment),
               # right_annotation = c(row_ha_ts,
               #                      #row_ha_regulation,
               #                      row_ha_chr,
               #                      gap = unit(1, "mm")),
               heatmap_width = unit(3.5, "cm"),
               heatmap_height = unit(10, "cm"),
               use_raster = TRUE,
               raster_device = "CairoPNG",
               raster_quality = 10,
               heatmap_legend_param = legends_param)
  
  return(h)
}
```

```{r draw_meth_reg}
fs <- 1.8
my_genes <- CT_genes %>% 
  filter(!is.na(DAC_induced))
h_met <- mean_methylation_fig(genes = my_genes$external_gene_name,
                              values_only = FALSE,
                              row_fontsize = fs,
                              column_fontsize = 5,
                              na.omit = FALSE)

h_DAC <- DAC_fig(my_genes$external_gene_name, 
                 multimapping = TRUE, 
                 units = "log1p",
                 row_fontsize = fs)


h_list <- h_met + h_DAC 
draw(h_list, main_heatmap = "Mean \nmethylation", ht_gap = unit(0.5, "cm"), 
     merge_legend = TRUE)
```

```{r, eval = FALSE, echo = FALSE}
pdf(file = paste0("figs/Methylation_classification.pdf"))
draw(h_list, main_heatmap = "Mean \nmethylation", ht_gap = unit(0.5, "cm"), 
     merge_legend = TRUE)
dev.off()
```

## Proportion of methylation regulated

```{r prop_meth}
x <- enframe(table(CT_genes_types %>% 
  filter(gene %in% CT_genes$external_gene_name) %>% 
  pull(Regulation)), name = "Regulation") %>% 
  left_join(enframe(prop.table(table(CT_genes_types$Regulation)), 
                    name = "Regulation", value = "Percent"))

met_prop_fig <- ggplot(x, aes(x = '', y = value, fill = Regulation)) +
  geom_bar(stat="identity", width=1) +
  coord_polar("y", start = 0) +
  geom_text(aes(y = value/5.5 + c(0, cumsum(value)[-length(value)]), 
                label = rev(paste0(Regulation, "\n", value, "/", 
                                   nrow(CT_genes_types %>%
                                          filter(!is.na(Regulation))), "\n",
                               round(Percent * 100), "%")))) +
  scale_fill_manual(values = Regulation_colors) +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.title = element_blank(),
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.ticks = element_blank(),
        axis.text.x = element_blank(),
        legend.position = "none")
met_prop_fig
```

```{r, eval = F}
pdf(file = paste0("figs/Met_proportions.pdf"), width = 4, height = 4)
met_prop_fig
dev.off()
```

## Enrichment on X chromosome

Genes regulated by methylation are enriched on the X chromosome.

```{r}
CT_genes_types$chr <- factor(CT_genes_types$chr, levels = c(1:22, "X", "Y"))

fig_X_enrichment <- ggplot(CT_genes_types %>% 
                             filter(!is.na(regulated_by_methylation)), 
                           aes(x = chr, fill = Regulation)) +
  geom_bar(stat = 'count') +
  xlab("Chromosome") +
  ylab("Gene number") +
  scale_fill_manual(values = Regulation_colors) +
  theme_bw() +
  theme(legend.position = "inside",
        legend.position.inside = c(0.2, 0.85),
        #aspect.ratio = 1.5,
        panel.grid = element_blank())

fig_X_enrichment
```


```{r, eval = FALSE, echo=FALSE}
pdf(file = paste0("figs/X_enrichment.pdf"), width = 4, height = 6)
fig_X_enrichment
dev.off()
```

Genes are enriched on the X chromosome 
(`r table(CT_genes$X_linked)[[2]]/dim(CT_genes)[1]*100`%). Also, 
`r table(CT_genes$regulated_by_methylation)[[2]]` genes have been flagged as 
regulated by methylation 
(`r table(CT_genes$regulated_by_methylation)[[2]]/dim(CT_genes)[1]*100`%). 
It is interesting to study the link between these two characteristics.

On the chromosome X, there is a clear enrichment of CT genes regulated by 
methylation (`r table(CT_genes$X_linked, CT_genes$regulated_by_methylation)[2, 2]`
/`r table(CT_genes$X_linked)[[2]]` chrX genes or 
`r table(CT_genes$X_linked, CT_genes$regulated_by_methylation)[2, 2]`/
`r table(CT_genes$regulated_by_methylation)[[2]]` genes regulated by 
methylation).

## Stats

### Testis-specific and methylation

CT genes are enriched among testis-specific genes regulated by methylation.
Proportion of Testis-specific genes regulated by methylation

```{r testis_spec_meth}
all_genes_chr <- all_genes %>% 
  mutate(chr = case_when(chr == "X" ~ "X-linked",
                           chr == "Y" ~ "Y-linked",
                           !chr %in% c("X", "Y") ~ "Autosome")) 

x <- table(met = all_genes_chr$regulated_by_methylation, 
           testis_specificity = all_genes_chr$testis_specificity, 
           useNA = "ifany")
x

chisq.test(x[c("FALSE", "TRUE"), c("not_testis_specific", "testis_specific")])
chisq.test(x[c("FALSE", "TRUE"), c("not_testis_specific",
                                   "testis_specific")])$p.value
```
- `r x["TRUE", "testis_specific"]`/`r x["TRUE", "testis_specific"] + x["FALSE", "testis_specific"]`
(`r x["TRUE", "testis_specific"]/(x["TRUE", "testis_specific"] + x["FALSE", "testis_specific"])*100`
%) of Testis-specific are regulated by methylation

=> Significant enrichment of genes regulated by methylation among testis-specific genes 
(pval = `r chisq.test(x[c("FALSE", "TRUE"), c("not_testis_specific", "testis_specific")])$p.value`)

Proportion of CT genes among Testis-specific genes regulated or not by methylation

```{r CT_meth}
TS_genes <- all_genes_chr %>% 
  filter(testis_specificity == "testis_specific")
x <- table(TS_genes = TS_genes$CT_gene_type, regulated_by_met = TS_genes$regulated_by_methylation)
x
chisq.test(x)
chisq.test(x)$p.value
```

- `r x["CT_gene", "TRUE"]` / `r x["CT_gene", "TRUE"] + x["other", "TRUE"]`
(`r x["CT_gene", "TRUE"]/(x["CT_gene", "TRUE"] + x["other", "TRUE"]) * 100`%)
of TS genes regulated by methylation are CT genes

- `r x["CT_gene", "FALSE"]` / `r x["CT_gene", "FALSE"] + x["other", "FALSE"]`
(`r x["CT_gene", "FALSE"]/(x["CT_gene", "FALSE"] + x["other", "FALSE"]) * 100`%)
of TS genes not regulated by methylation are CT genes

=> Strong enrichment in CT genes among Testis-specific genes regulated by 
methylation (`chisq.test(x)$p.value`)

```{r}
fig_TS_CT_genes <- TS_genes %>% 
  filter(!is.na(regulated_by_methylation)) %>% 
  mutate(Regulation = case_when(regulated_by_methylation ~ "Met",
                                !regulated_by_methylation ~ "Not Met")) %>% 
  mutate(Regulation = factor(Regulation, levels = c("Not Met", "Met"))) %>% 
  mutate(Group = case_when(CT_gene_type == "CT_gene" ~ "CT gene",
                             CT_gene_type != "CT_gene" ~ "Not CT gene")) %>% 
  group_by(Regulation, Group) %>% 
  dplyr::count() %>% 
  mutate(testis_specificity = "Testis-specific genes") %>% 
  ggplot(aes(x = Regulation, y = n, fill = Group)) +
  geom_col() +
  scale_fill_manual(values = c("seagreen2", "gray50")) +
  xlab('') +
  ylab("Number of genes") +
  facet_wrap(~ testis_specificity) +
  theme_bw() +
  theme(legend.position = "right",
        plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank(),
        strip.text = element_text(size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10))        
fig_TS_CT_genes

```

### Chromosomal location

CT chromosomal location

```{r}
x <- table(gene = all_genes_chr$CT_gene_type, chr = all_genes_chr$chr)
x
```

- `r x["CT_gene", "X-linked"]` / `r sum(x["CT_gene", ])`
(`r x["CT_gene", "X-linked"]/sum(x["CT_gene", ]) * 100`%) CT genes are X-linked 

```{r}
CT_genes_chr <- all_genes_chr %>% 
  filter(CT_gene_type == "CT_gene")

x <- table(chr = CT_genes_chr$chr, 
           regulated_by_met = CT_genes_chr$regulated_by_methylation)
x
```

- `r x["X-linked", "TRUE"]` / `r sum(x["X-linked", ])`
(`r x["X-linked", "TRUE"]/sum(x["X-linked", ]) * 100`%)
of X-linked genes are regulated by methylation
 
```{r}
chisq.test(x[c("Autosome", "X-linked"), c("FALSE", "TRUE")])$p.value
```

All genes Chromosomal location

```{r}
x <- table(chr = all_genes_chr$chr, 
           testis_specificity = all_genes_chr$testis_specificity)
x

chisq.test(x[c("Autosome", "X-linked"), c("not_testis_specific", 
                                          "testis_specific")])

chisq.test(x[c("Autosome", "X-linked"), c("not_testis_specific", 
                                          "testis_specific")])$p.value

chisq.test(x[c("Autosome", "Y-linked"), c("not_testis_specific",
                                          "testis_specific")])$p.value
```

- Testis-specific genes are enriched on the X (pval = 
`r chisq.test(x[c("Autosome", "X-linked"), c("not_testis_specific", "testis_specific")])$p.value`)

- Testis-specific genes are also enriched on the Y (pval = 
`r chisq.test(x[c("Autosome", "Y-linked"), c("not_testis_specific", "testis_specific")])$p.value`)

- But testis-preferential genes are not enriched on the X (pval 
`r chisq.test(x[c("Autosome", "X-linked"), c("not_testis_specific", "testis_preferential")])$p.value`)

- and testis-preferential genes are not enriched on the X (pval 
`r chisq.test(x[c("Autosome", "Y-linked"), c("not_testis_specific", "testis_preferential")])$p.value`)


```{r}
chisq.test(x[c("Autosome", "X-linked"), c("not_testis_specific", "testis_preferential")])

chisq.test(x[c("Autosome", "Y-linked"), c("not_testis_specific", "testis_preferential")])
```

### Chromosomal location, testis specificity and regulation by methylation 

Proportion of X-linked genes among Testis-specific genes regulated or not by 
methylation.

```{r}
TS_genes_chr <- all_genes_chr %>% 
  filter(testis_specificity == "testis_specific")
x <- table(chr = TS_genes_chr$chr, 
           TS_genes_regulated_by_met = TS_genes_chr$regulated_by_methylation, 
           useNA = "ifany")
x
```

- `r x["X-linked", "TRUE"]` / `r sum(x[, "TRUE"])`
(`r x["X-linked", "TRUE"]/sum(x[, "TRUE"]) * 100`%) of testis-specific genes 
regulated by methylation are X-linked

- `r x["X-linked", "FALSE"]` / `r sum(x[, "FALSE"])`
(`r x["X-linked", "FALSE"]/sum(x[, "FALSE"]) * 100`%) of testis-specific genes 
 not regulated by methylatoin are X-linked

```{r}
chisq.test(x[c("Autosome", "X-linked"), c("FALSE", "TRUE")])
chisq.test(x[c("Autosome", "X-linked"), c("FALSE", "TRUE")])$p.value
```

```{r}
fig_TS_X <- TS_genes_chr %>% 
  filter(!is.na(regulated_by_methylation)) %>% 
  mutate(chr = case_when(chr == "X-linked" ~ "chrX",
                         chr == "Y-linked" ~ "chrY",
                         !chr %in% c("X-linked", "Y-linked") ~ "Autosome")) %>%
  mutate(Regulation = case_when(
    regulated_by_methylation ~ 
      "Testis-specific genes\nregulated by methylation",
    !regulated_by_methylation ~ 
      "Testis-specific genes\nNOT regulated by methylation")) %>% 
  mutate(Regulation = 
           factor(Regulation, 
                  levels = 
                    c("Testis-specific genes\nNOT regulated by methylation", 
                      "Testis-specific genes\nregulated by methylation"))) %>% 
  mutate(Group = case_when(CT_gene_type == "CT_gene" ~ "CT gene",
                           CT_gene_type != "CT_gene" ~ "Not CT gene")) %>% 
  group_by(Regulation, chr, Group) %>% 
  dplyr::count() %>% 
  ggplot(aes(x = chr, y = n, fill = Group)) +
  geom_col() + #width = 0.5) +
  scale_fill_manual(values = c("seagreen2", "gray50")) +
  facet_wrap(~ Regulation, scale = "free_y") +
  xlab('') +
  ylab("") +
  theme_bw() +
  theme(legend.position = "right",
        plot.title = element_text(hjust = 0.5),
        #title = element_text(size = 10, face = "bold",
        #aspect.ratio = 1.5,
        panel.grid = element_blank(),
        strip.text = element_text(size = 9),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 12))


fig_TS_X
```



## Put figs together

```{r}
fig <- cowplot::plot_grid(fig_TS_CT_genes + theme(legend.position = "none"), 
                          fig_TS_X + theme(legend.position = "none"),
           rel_widths = c(1,2))
          

legend <- cowplot::get_legend(fig_TS_CT_genes)

cowplot::plot_grid(fig, legend, align = "v", rel_widths = c(1, 0.2))
```

```{r}
pdf(file = paste0("figs/Testis_specific.pdf"), width = 8, height = 4)
cowplot::plot_grid(fig, legend, 
                   rel_widths = c(1, 0.2))
  
dev.off()
```


Putting the figs on the same panel with a bit of fine tuning

```{r meth_figs}
x <- enframe(table(CT_genes_types$Regulation), name = "Regulation") %>% 
  left_join(enframe(prop.table(table(CT_genes_types$Regulation)), 
                    name = "Regulation", value = "Percent"))

prop_fig <- ggplot(x, aes(x = '', y = value, fill = Regulation)) +
  geom_bar(stat="identity", width = 1) +
  coord_polar("y", start = 0) +
  geom_text(aes(y = value/5 + c(0, cumsum(value)[-length(value)]),
                label = rev(paste0(Regulation, "\n", value, "/", 
                                   sum(table(CT_genes_types$Regulation)), "\n",
                               round(Percent * 100), "%"))), size = 1.2 )+
  scale_fill_manual(values = Regulation_colors) +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.title = element_blank(),
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.ticks = element_blank(),
        axis.text.x = element_blank(),
        legend.position = "none")


fig_freq_by_chr <- ggplot(CT_genes_types[
  !is.na(CT_genes_types$regulated_by_methylation), ], 
  aes(x = chr, fill = Regulation)) +
  geom_bar(stat = 'count') +
  xlab("Chromosome") +
  ylab("Gene number") +
  scale_fill_manual(values = Regulation_colors) +
  theme_test()+
  theme(legend.position = "top",
        aspect.ratio = 1,
        axis.text = element_text(size = 3),
        axis.title = element_text(size = 5),
        axis.ticks = element_line(linewidth = 0.2),
        axis.ticks.length = unit(0.5, "mm"),
        legend.text = element_text(size = 5),
        legend.key.size = unit(2, "mm"),
        legend.title = element_blank())


right_fig <- cowplot::plot_grid(prop_fig, fig_freq_by_chr, fig,
                                labels = c('B', 'C', 'D'), 
                                #hjust = -2, 
                                vjust = 1,
                                #scale = c(0.8, 1), 
                                rel_heights = c(0.5, 1),
                                ncol = 1)

left_fig <- cowplot::plot_grid(grid.grabExpr(
  draw(h_list, main_heatmap = "Mean \nmethylation", ht_gap = unit(0.3, "cm"), 
       merge_legend = TRUE)), labels = c('A'))


cowplot::plot_grid(left_fig, right_fig, 
          rel_widths = c(2,1),
          #scale = c(0.6, 0.5),
          label_size = 10, nrow = 1, 
          #align = "h", 
          align = "vh")
```


```{r, eval = FALSE, echo = FALSE}
pdf(file = paste0("figs/Methylation_panel.pdf"), width = 7, height = 4)
cowplot::plot_grid(left_fig, right_fig, 
          rel_widths = c(2,1),
          scale = c(1, 0.8),
          label_size = 10, nrow = 1,
          align = "h")
dev.off()
```
# X-linked are more testis-specific than autosomal genes

```{r}
TS_and_TP_genes_chr <- all_genes_chr %>% 
  filter(CT_gene_type %in% c("CT_gene", "CTP_gene"))
table(chr = TS_and_TP_genes_chr$chr, 
      TS_and_TP_genes = TS_and_TP_genes_chr$testis_specificity)
```



# Gene biotype

Mainly protein-coding genes
Most genes are mainly protein coding genes 
(`r table(CT_genes$transcript_biotype)["protein_coding"]/dim(CT_genes)[1]*100`%).

```{r}
table(Gene = CT_genes_chr$chr, CT_gene_biotype = CT_genes_chr$transcript_biotype)
```
```{r}
table(Gene = TS_genes_chr$chr, Testis_specific_genes = TS_genes_chr$transcript_biotype)
```

Gene biotype / regulation

```{r}
table(CT_genes_chr$transcript_biotype,
                 methylation = CT_genes_chr$regulated_by_methylation)
```


# CT genes in tumour : coexpression and link with global hypomethylation

## Positive tumour proportion and expression level

CT genes regulated by methylation are activated more frequently and at higher 
levels

```{r pos_tum_exprs}
TCGA <- TCGA_TPM
TCGA <- TCGA[, TCGA$shortLetterCode != "NT"]

tumors <- unique(TCGA$project_id)

thr_pos <- 1

prop_of_pos_tum <- as_tibble(assay(TCGA), rownames = "ensembl_gene_id") %>% 
  filter(ensembl_gene_id %in% CT_genes$ensembl_gene_id) %>% 
  pivot_longer(names_to = "barcode", values_to = "TPM", -ensembl_gene_id) %>% 
  mutate(activated = ifelse(TPM >= thr_pos, "n_pos", "n_neg")) %>% 
  group_by(ensembl_gene_id, activated) %>% 
  dplyr::summarize(n = n()) %>% 
  pivot_wider(names_from = activated, values_from = n) %>% 
  mutate(proportion_of_positive_tum = n_pos * 100 / (n_pos + n_neg)) 

mean_in_pos <- as_tibble(assay(TCGA), rownames = "ensembl_gene_id") %>% 
  filter(ensembl_gene_id %in% CT_genes$ensembl_gene_id) %>%  
  pivot_longer(names_to = "barcode", values_to = "TPM", -ensembl_gene_id) %>% 
  mutate(activated = ifelse(TPM >= thr_pos, "n_pos", "n_neg")) %>% 
  filter(activated == "n_pos") %>% 
  group_by(ensembl_gene_id) %>% 
  summarise(mean_log = log1p(mean(TPM)),
            mean_TPM = mean(TPM)) 

mean_and_prop_TCGA <- prop_of_pos_tum %>% 
  left_join(mean_in_pos) %>% 
  right_join(CT_genes_types) %>% 
  dplyr::select(gene, mean_log, proportion_of_positive_tum, Regulation, 
                X_linked) %>% 
  mutate(source = "TCGA tumor samples")

# combine with CCLE data
load("~/cluster/Packages/CTdata/eh_data/CCLE_data.rda")
ccle <- CCLE_data
tumors <- unique(ccle$primary_disease)

prop_of_pos_tum <- as_tibble(assay(ccle), rownames = "ensembl_gene_id") %>% 
  filter(ensembl_gene_id %in% CT_genes$ensembl_gene_id) %>%  
  pivot_longer(names_to = "cell", values_to = "TPM", -ensembl_gene_id) %>% 
  mutate(activated = ifelse(TPM >= thr_pos, "n_pos", "n_neg")) %>% 
  group_by(ensembl_gene_id, activated) %>% 
  dplyr::summarize(n = n()) %>% 
  pivot_wider(names_from = activated, values_from = n) %>% 
  mutate(proportion_of_positive_tum = n_pos * 100 / (n_pos + n_neg)) 

mean_in_pos <- as_tibble(assay(ccle), rownames = "ensembl_gene_id") %>% 
  filter(ensembl_gene_id %in% CT_genes$ensembl_gene_id) %>%   
  pivot_longer(names_to = "cell", values_to = "TPM", -ensembl_gene_id) %>% 
  mutate(activated = ifelse(TPM >= thr_pos, "n_pos", "n_neg")) %>% 
  filter(activated == "n_pos") %>% 
  group_by(ensembl_gene_id) %>% 
  summarise(mean_log = log1p(mean(TPM)),
            mean_TPM = mean(TPM)) 

mean_and_prop_CCLE <- prop_of_pos_tum %>% 
  left_join(mean_in_pos) %>% 
  right_join(CT_genes_types) %>% 
  dplyr::select(gene, mean_log, proportion_of_positive_tum, Regulation, 
                X_linked) %>% 
  mutate(source = "CCLE tumor cell lines")

mean_and_prop <- rbind(mean_and_prop_TCGA, mean_and_prop_CCLE) %>% 
  filter(!is.na(Regulation))


fig_mean_prop_in_tum_horizontal <- 
  ggplot(mean_and_prop,
         aes(x = mean_log, y = proportion_of_positive_tum,
             shape = X_linked,
             color = Regulation,
             label = gene)) +
  geom_point(size = 1) +
  scale_colour_manual(values = Regulation_colors)+
  ggrepel::geom_text_repel(
    data = mean_and_prop %>% 
      filter(proportion_of_positive_tum > 20 | 
               (proportion_of_positive_tum > 10 & mean_log > 3.5)), 
    aes(x = mean_log, y = proportion_of_positive_tum, 
        label = gene), size = 1.5, color = "black", min.segment.length = 4,
    max.overlaps = 30, direction = "both") +
  xlab("Mean expression log(TPM + 1) in positive tumors") +
  ylab("Proportion of positive tumors (%)") +
  facet_wrap(~ source, ncol = 2) +
  theme_bw() +
  theme(legend.position = "right",
        aspect.ratio = 0.8,
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        legend.key.size = unit(0.3, 'cm'),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 8),
        axis.text = element_text(size = 7),
        axis.ticks = element_line(size = 0.1),
        axis.title = element_text(size = 8),
        strip.text = element_text(size = 8)) +
  ylim(0,42)

fig_mean_prop_in_tum_horizontal
```
```{r, eval = FALSE, echo = FALSE}
pdf(file = paste0("figs/Mean_prop_in_tumors.pdf"), width = 6, height = 3)
fig_mean_prop_in_tum_horizontal
dev.off()
```

## Activation in TCGA

```{r}
# slightly modified the function 
print_CCLE_heatmap <- function(genes = NULL,
                               tum_type = c("Skin", "Lung", 
                                            "Colorectal", 
                                            "Breast", 
                                            "Head_and_Neck",
                                            "Esophageal"),
                               h_width = 5,
                               h_height = 10,
                               downsample_to_n = 450,
                               fontsize = 4, 
                               clust_method = "ward.D",
                               split_rows_by = "regulation", 
                               show_annotation_legend = TRUE,
                               clustRow = FALSE,
                               clustCol = TRUE){
  
  ## downsample tumors to have the same number of tumors in each type
  ccle <- CCLE_data
  CCLE <- ccle[, ccle$type == tum_type[1]]
  CCLE <- CCLE[, 1:min(ncol(CCLE), downsample_to_n)]
  
  for (tum in tum_type[-1]){
    tmp <- ccle[, ccle$type == tum]
    tmp <- tmp[, 1:min(ncol(tmp), downsample_to_n)]
    CCLE <- cbind(CCLE, tmp)
  }
  
  split_by <- factor(CCLE$type, levels = tum_type)
  
  column_ha_tumor <- HeatmapAnnotation(
    CCLE_cell_type = CCLE$type,
   # border = TRUE,
    col = list(CCLE_cell_type = CCLE_colors),
    annotation_label = gpar(fontsize = 4),
    annotation_name_gp = gpar(fontsize = 0),
    simple_anno_size = unit(0.1, "cm"),
    annotation_legend_param = legends_param)
  
  ## Use gene names instead of ENSEMBL IDs
  mat <- SummarizedExperiment::assay(CCLE)
  rownames(mat) <- rowData(CCLE)$external_gene_name
  mat <- log1p(mat)
  
    
  gene_type <- all_genes_clean %>% 
    filter(gene %in% genes) %>% 
    mutate(gene = factor(gene, levels = genes)) %>% 
    arrange(gene)
   
 rowSplit <- NULL
  if(!is.null(split_rows_by)){
    if (split_rows_by == "regulation") {
      if (clustRow) gene_type <- gene_type %>% arrange(Regulation)
      rowSplit <- gene_type$Regulation
    }
    if (split_rows_by == "testis_specificity") {
      if (clustRow) gene_type <- gene_type %>% arrange(testis_specificity)
      rowSplit <- gene_type$testis_specificity
    }
    if (split_rows_by == "GTEX_specificity") {
      if (clustRow) gene_type <- gene_type %>% arrange(GTEX_specificity)
      rowSplit <- gene_type$GTEX_specificity
    }
    
    if (split_rows_by == "chr_met") {
      if (clustRow) gene_type <- gene_type %>% arrange(chr_met)
      rowSplit <- gene_type$chr_met
    }
  }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(Regulation = Regulation,
                                     annotation_legend_param = legends_param,
                                     simple_anno_size = unit(0.2, "cm"),
                                     col = list(Regulation = Regulation_colors),
                                     annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(chr_met = chr_met,
                                  annotation_legend_param = legends_param,
                                  simple_anno_size = unit(0.2, "cm"),
                                  col = list(chr_met = chr_met_colors),
                                  annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$Chr)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              simple_anno_size = unit(0.2, "cm"),
                              col = list(Chr = Chr_colors),
                              annotation_name_gp = gpar(fontsize = 0))
  
  GTEX_specificity <- as.matrix(gene_type$GTEX_specificity)
  rownames(GTEX_specificity) <- gene_type$gene
  row_ha_GTEX <- rowAnnotation(GTEX_specificity = GTEX_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(GTEX_specificity = 
                                          GTEX_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  testis_specificity <- as.matrix(gene_type$testis_specificity)
  rownames(testis_specificity) <- gene_type$gene
  row_ha_TS <- rowAnnotation(specificity = testis_specificity,
                             annotation_legend_param = legends_param,
                             simple_anno_size = unit(0.2, "cm"),
                             col = list(testis_specificity = 
                                          Testis_specificity_colors),
                             annotation_name_gp = gpar(fontsize = 0))
  
  rAnnot <- NULL
  if(show_annotation_legend){
    if (split_rows_by == "regulation") {
  rAnnot <- c(row_ha_regulation, gap = unit(1, "mm"))
    }
    if (split_rows_by == "testis_specificity") {
  rAnnot <- c(row_ha_TS, gap = unit(1, "mm"))
    }
    if (split_rows_by == "GTEX_specificity") {
  rAnnot <- c(row_ha_GTEX, gap = unit(1, "mm"))
    }
    
    if (split_rows_by == "chr_met") {
  rAnnot <- c(row_ha_TS, gap = unit(1, "mm"))
    }
  }

  
  # heatmap_colors <- colorRamp2(
  #   seq(0, quantile(rowMax(mat), 0.8), length = 11),                 
  #   legend_colors)
  
  heatmap_colors <- colorRamp2(
    seq(0, 5, length = 11),                 
    legend_colors)
  
  Heatmap(mat[as.vector(gene_type$gene), ],
          name = "CCLE (logTPM)",
          #column_title = paste0("CCLE tumor cell lines"),
          column_title_gp = gpar(fontsize = 2, face = "bold"),
          clustering_method_rows = clust_method,
          clustering_method_columns = clust_method,
          column_split = split_by,
          row_gap = unit(c(0.5), "mm"),
          col = heatmap_colors,
          cluster_rows = clustRow,
          cluster_columns = clustCol,
          cluster_column_slices = FALSE,
          show_row_names = TRUE,
          show_column_names = FALSE,
          show_row_dend = FALSE,
          show_column_dend = FALSE,
          column_names_gp = gpar(fontsize = 6),
          column_names_centered = TRUE,
          row_names_gp = gpar(fontsize = fontsize),
          row_names_side = "left",
          border = TRUE,
          border_gp = gpar(lwd = 0.1),
          column_names_side = c("top"),
          column_names_rot = 90,
          row_title_gp = gpar(fontsize = 0),
          #top_annotation = column_ha_tumor,
          right_annotation = rAnnot,
          row_split = rowSplit,
          width = unit(h_width, "cm"),
          height = unit(h_height, "cm"),
          use_raster = TRUE,
          raster_device = "CairoPNG",
          raster_quality = 10,
          heatmap_legend_param = legends_param,
          show_heatmap_legend = TRUE)
}
```

```{r}
met_genes <- CT_genes_types %>% 
  filter(regulated_by_methylation) %>% 
  pull(gene)

h1 <- print_CCLE_heatmap(met_genes, 
                         downsample_to_n = 100, 
                         h_width = 2.5,
                         h_height = 2.5,
                         fontsize = .5, 
                         tum_type = c("Skin", "Lung", 
                                      "Brain", "Breast","Lymphoma", 
                                      "Leukemia", "Colorectal"),
                   clust_method = "ward.D",
                   clustRow = TRUE,
                  # show_row_names = TRUE,
                  # show_top_annotations = FALSE,
                   show_annotation_legend = FALSE,
                   split_rows_by = "regulation")
draw(h1, column_title = "CT genes regulated by methylation",
     column_title_gp = gpar(fontsize = 2.5), show_heatmap_legend = FALSE,
     show_annotation_legend = FALSE)

not_met_genes <- CT_genes_types %>% 
  filter(!regulated_by_methylation) %>%
  pull(gene)


h2 <- print_CCLE_heatmap(not_met_genes, 
                         fontsize = 0.5, 
                         downsample_to_n = 100, 
                         h_width = 2.5,
                         h_height = 1,
                         tum_type = c("Skin", "Lung", 
                                      "Brain", "Breast","Lymphoma", 
                                      "Leukemia", "Colorectal"),
                         clust_method = "ward.D",
                         clustRow = TRUE,
                         show_annotation_legend = F,
                         split_rows_by = "regulation")

draw(h2, column_title = "CT genes not regulated by methylation",
     column_title_gp = gpar(fontsize = 2.5), show_heatmap_legend = TRUE,
     heatmap_legend_side = "bottom",
     show_annotation_legend = FALSE)

```

```{r, eval = FALSE, echo = FALSE}
pdf(file = paste0("figs/CT_met_CCLE.pdf"), width = 1.3, height = 1.5)
draw(h1, column_title = "CT genes regulated by methylation",
     column_title_gp = gpar(fontsize = 2.5), show_heatmap_legend = FALSE,
     show_annotation_legend = FALSE)
dev.off()

pdf(file = "figs/CT_not_met_CCLE.pdf", height = 2, width = 1.5)
draw(h2, column_title = "CT genes not regulated by methylation",
     column_title_gp = gpar(fontsize = 2.5), show_heatmap_legend = TRUE,
     heatmap_legend_side = "bottom",
     show_annotation_legend = FALSE)
dev.off()
```

# Oncogenes and Tumor suppressor

```{r}
onco_TS_list <- list("Tumor suppressors" = CT_genes %>% 
                       filter(tumor_suppressor == "tumor_suppressor") %>%
                       pull(external_gene_name),
                     Oncogenes = CT_genes %>% 
                       filter(oncogene == "oncogene") %>%
                       pull(external_gene_name))
onco_TS_list

gp <- VennThemes(compute.Venn(Venn(onco_TS_list)))
gp[["Face"]][["10"]]$fill <-  "lightsteelblue1"
gp[["Face"]][["01"]]$fill <-  "mediumaquamarine"
gp[["Face"]][["11"]]$fill <-  "gray80"
gp[["Set"]][["Set2"]]$col <-  "darkseagreen4"
gp[["Set"]][["Set1"]]$col <-  "lightsteelblue4"
gp[["SetText"]][["Set1"]]$col <-  "steelblue"
gp[["SetText"]][["Set2"]]$col <-  "paleturquoise4"
gp[["FaceText"]][["10"]]$col <-  "black"
plot(Vennerable::Venn(onco_TS_list), gp = gp)
```

```{r, eval = FALSE, echo = FALSE}
pdf(file = paste0("figs/Oncogenic_CT_Venn.pdf"))
plot(Venn(onco_TS_list), gp = gp)
dev.off()
```

# CT expression in healthy cells (fetal and adult)

## Heatmap

First treating data

```{r}
# Missing genes are probably just not detected
# Add zero counts to  missing genes in all datasets

testis_sce <- testis_sce[CT_genes$external_gene_name[
  CT_genes$external_gene_name %in% rownames(testis_sce)],]


# add missing CT genes (not in the downloaded dataset, as not detected 
# (below detection) => were removed)
# Add zero count values for these genes
rowData(testis_sce) <- NULL
missing_CT <- CT_genes$external_gene_name[!CT_genes$external_gene_name %in% 
                                            rownames(testis_sce)]
missing_CT
sce_missing_CT <- testis_sce[1:length(missing_CT),]
rownames(sce_missing_CT) <- missing_CT
count_mat_missing_CT <- matrix(nrow = length(missing_CT), 
                               ncol = ncol(testis_sce), data = 0)
rownames(count_mat_missing_CT) <- missing_CT
colnames(count_mat_missing_CT) <- colnames(testis_sce)
assay(sce_missing_CT,1) <- count_mat_missing_CT
assay(sce_missing_CT,2) <- count_mat_missing_CT
testis_sce <- rbind(testis_sce, sce_missing_CT)


# add missing CT genes (not in the downloaded dataset, as not detected 
# (below detection) => were removed)
# Add zero count values for these genes
rowData(FGC_sce) <- NULL

FGC_sce <- FGC_sce[CT_genes$external_gene_name[
  CT_genes$external_gene_name %in% rownames(FGC_sce)],]

missing_CT <- CT_genes$external_gene_name[!CT_genes$external_gene_name %in%
                                            rownames(FGC_sce)]
sce_missing_CT <- FGC_sce[1:length(missing_CT),]
rownames(sce_missing_CT) <- missing_CT
count_mat_missing_CT <- matrix(nrow = length(missing_CT), 
                               ncol = ncol(FGC_sce), data = 0)
rownames(count_mat_missing_CT) <- missing_CT
colnames(count_mat_missing_CT) <- colnames(FGC_sce)
assay(sce_missing_CT,1) <- count_mat_missing_CT
assay(sce_missing_CT,2) <- count_mat_missing_CT
FGC_sce <- rbind(FGC_sce, sce_missing_CT)


# add missing CT genes (not in the downloaded dataset, as not detected (below 
# detection) => were removed)
# Add zero count values for these genes
rowData(oocytes_sce) <- NULL
oocytes_sce <- oocytes_sce[CT_genes$external_gene_name[
  CT_genes$external_gene_name %in% rownames(oocytes_sce)],]
missing_CT <- CT_genes$external_gene_name[!CT_genes$external_gene_name %in% 
                                            rownames(oocytes_sce)]
sce_missing_CT <- oocytes_sce[1:length(missing_CT),]
rownames(sce_missing_CT) <- missing_CT
count_mat_missing_CT <- matrix(nrow = length(missing_CT), 
                               ncol = ncol(oocytes_sce), data = 0)
rownames(count_mat_missing_CT) <- missing_CT
colnames(count_mat_missing_CT) <- colnames(oocytes_sce)
assay(sce_missing_CT,1) <- count_mat_missing_CT
assay(sce_missing_CT,2) <- count_mat_missing_CT
oocytes_sce <- rbind(oocytes_sce, sce_missing_CT)



# pool Sperm1 and Sperm2 types
levels(testis_sce$type)[c(7,8)] <- "Sperm"
germcells <-  c("SSC", "Spermatogonia", "Early_spermatocyte", 
                "Late_spermatocyte", "Round_spermatid", 
                "Elongated_spermatid", "Sperm")

# add meiotic stage to the colData of testis_sce
testis_sce$stage <- "pre-meiotic"
testis_sce[, testis_sce$type %in% 
             c("Early_spermatocyte", "Late_spermatocyte")]$stage <- "meiotic"
testis_sce[, testis_sce$type %in% 
             c("Round_spermatid", "Elongated_spermatid", "Sperm")]$stage <- 
  "post-meiotic"
testis_sce[, testis_sce$type %in% 
             c("Macrophage", "Endothelial", "Myoid", "Sertoli", 
               "Leydig")]$stage <- "somatic"
testis_sce$stage <- factor(
  testis_sce$stage, 
  levels = c("pre-meiotic", "meiotic", "post-meiotic", "somatic"))


# Remove genes with rowcounts = 0 in all datasets
# !!! Few genes are not detetected in testis_Sce but in FGC or oocytes
not_in_testis_sce <- rownames(testis_sce)[rowSums(assay(testis_sce)) == 0]
not_in_FGC_sce <- rownames(FGC_sce)[rowSums(assay(FGC_sce)) == 0]
not_in_oocytes_sce <- rownames(oocytes_sce)[rowSums(assay(oocytes_sce)) == 0]
not_in_testis_FGC <- BiocGenerics::intersect(not_in_testis_sce, not_in_FGC_sce)
not_in_testis_FGC_oocytes <- BiocGenerics::intersect(not_in_testis_FGC, 
                                                     not_in_oocytes_sce)

kept_genes <- CT_genes %>% 
  filter(!is.na(regulated_by_methylation)) %>% 
  filter(!external_gene_name %in% not_in_testis_FGC_oocytes) %>% 
  pull(external_gene_name)

never <- CT_genes %>% 
  filter(!is.na(regulated_by_methylation)) %>% 
  filter(external_gene_name %in% not_in_testis_FGC_oocytes) %>% 
  pull(external_gene_name)

legends_param <- list(
  labels_gp = gpar(col = "black", fontsize = 2),
  title_gp = gpar(col = "black", fontsize = 3),
  simple_anno_size = unit(0.15, "cm"),
  row_names_gp = gpar(fontsize = 1),
  annotation_name_side = "left",
  border = FALSE,
  border_gp = gpar(lwd = 0.01),
  grid_width = unit(0.2, "cm"),
  grid_height = unit(0.2, "cm"),
  legend_height = unit(0.1, "cm"),
  annotation_name_gp = gpar(fontsize = 0))

# To label Y linked genes on the heatmaps
CT_genes_types <- CT_genes_types %>% 
  mutate(chromosome = case_when(chr == "X" ~ "X-linked",
                                chr == "Y" ~ "Y-linked",
                                !chr %in% c("X", "Y") ~ "Autosome"))
table(CT_genes_types$chromosome)
```
Kept `r length(kept_genes)` CT genes for the analysis:

CT genes with logcounts >= 0 in at least  one of the dataset

```{r healthy_cells_function}

# Function to downsample cell types

sample_cells <- function(barcodes, nmax){
  sample(barcodes, size = min(nmax, length(barcodes)), replace = FALSE)
}


## Function to draw expression in FGC with specific annotations
## and downasampling the number of cells from each cell type

draw_FGC_heatmap <- function(genes, 
                             split_by = "chr_met", 
                             ncells_max = NULL,
                             cell_type = NULL,
                             scale_lims = NULL,
                             sex = c("F", "M"),
                             clusterRow = FALSE, 
                             font_size = 6,
                             show_legend = TRUE,
                             h_width = 5,
                             h_height = 10,
                             show_right_annot = TRUE){
  
  fetal_sce <- FGC_sce
  
  if (!is.null(cell_type)) fetal_sce <- fetal_sce[,FGC_sce$type %in% cell_type]
  
  # Set the colors used in the final heatmap based on all samples (F + M), 
  # even when representing only F or M samples on the heatmap
  # => same scale of colors
  mat <-  as.matrix(assay(fetal_sce, "logcounts")[genes, ])
  
  if (is.null(scale_lims)) {
    scale_lims <- c(0, quantile(rowMaxs(mat), 0.8))
  }
  
  heatmap_colors <- colorRamp2(
    seq(scale_lims[1], scale_lims[2], length = 11),                 
    legend_colors)
  
  fetal_sce <- fetal_sce[, fetal_sce$sex %in% sex]
  
  df_col_FCS <- data.frame(development_stage = fetal_sce$development_stage,
                           sex = fetal_sce$sex,
                           stage = fetal_sce$stage,
                           donor_id = fetal_sce$donor_id,
                           fetal_cell_type = fetal_sce$cell_type,
                           type = fetal_sce$type)
  
  rownames(df_col_FCS) <- colnames(fetal_sce)
  df_col_FCS <- df_col_FCS[order(df_col_FCS$type),]
  
  #Randomly pick ncells_max cells of each cell type
  if (!is.null(ncells_max)) {
    set.seed(123)
    selected_cells <- sample_cells(
      rownames(df_col_FCS[df_col_FCS$type == levels(df_col_FCS$type)[1],]), 
      ncells_max)
    for (type in levels(df_col_FCS$type)[-1]){
      tmp <- sample_cells(
        rownames(df_col_FCS[df_col_FCS$type == type,]), ncells_max)
      selected_cells <- c(selected_cells, tmp)
    }
    df_col_FCS <- df_col_FCS[selected_cells,]
  }
  
  column_ha_type = HeatmapAnnotation(
    Fetal_cell_type = df_col_FCS$type,
    col = list(Fetal_cell_type = Fetal_cell_colors),   
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_development_stage = HeatmapAnnotation(
    development_stage = df_col_FCS$development_stage,
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_sex = HeatmapAnnotation(
    sex = df_col_FCS$sex,
    col = list(sex = sex_colors),
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_donor_id = HeatmapAnnotation(
    donor_id = df_col_FCS$donor_id,
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_stage = HeatmapAnnotation( 
    stage = df_col_FCS$stage,
    col = list(stage = stage_colors),
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) 
  
  if (split_by == "regulation") {
    gene_type <- gene_type %>% arrange(Regulation)
    rowSplit <- gene_type$Regulation
  }
  if (split_by == "testis_specificity") {
    gene_type <- gene_type %>% arrange(Testis_specificity)
    rowSplit <- gene_type$Testis_specificity
  }
  if (split_by == "chr_met") {
    gene_type <- gene_type %>% arrange(chr_met)
    rowSplit <- gene_type$chr_met
  }
  if (!clusterRow & split_by == "chr_met") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$chr_met
  }
  if (split_by == "met_3_groups") {
    gene_type <- gene_type %>% arrange(met_3_groups)
    rowSplit <- gene_type$met_3_groups
  }
   if (!clusterRow & split_by == "met_3_groups") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$met_3_groups
   }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(
    Regulation = Regulation,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(Regulation = Regulation_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(
    chr_met = chr_met,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(chr_met = chr_met_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$chromosome)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              border = FALSE,
                              simple_anno_size = unit(0.15, "cm"), 
                              col = list(Chr = Chr_colors_with_Y),
                              annotation_name_gp = gpar(fontsize = 0))
  
  mat <- as.matrix(
    assay(fetal_sce, "logcounts")[gene_type$gene, rownames(df_col_FCS)])
  
  rAnnot <- c(row_ha_regulation, row_ha_chr, gap = unit(1, "mm"))
  
  if (!show_right_annot) rAnnot <- NULL
  
  title <- "Fetal germ cells"
  if (all(sex == "F")) {
    title <- "Female fetal germ cells"
  }
  if (all(sex == "M")) {
    title <- "Male fetal germ cells"
  }
  
  if (is.null(scale_lims)) {
    scale_lims <- c(0, quantile(rowMaxs(mat), 0.8))
  }
  
  heatmap_colors <- colorRamp2(
    seq(scale_lims[1], scale_lims[2], length = 11),                 
    legend_colors)
  
  ht_FGCs <- Heatmap(mat,
                     name = "Logcounts",
                     column_title = title,
                     column_split = df_col_FCS$type,
                     cluster_row_slices = FALSE,
                     cluster_column_slices = FALSE,
                     col = heatmap_colors,
                     border = TRUE,
                     border_gp = gpar(lwd = 0.2),
                     row_gap = unit(c(1), "mm"),
                     column_gap = unit(c(0.5), "mm"),
                     show_column_dend = FALSE,
                     show_row_dend = FALSE,
                     row_names_gp = gpar(fontsize = font_size),
                     row_names_side = "left",
                     cluster_rows = clusterRow,
                     cluster_columns = FALSE,
                     show_column_names = FALSE,
                     show_row_names = TRUE,
                     row_split = rowSplit,
                     top_annotation = c(column_ha_sex, 
                                        column_ha_stage,
                                        #column_ha_group, 
                                        #column_ha_donor_id,
                                        column_ha_type),  
                     right_annotation = rAnnot,
                     column_title_gp = gpar(fontsize = 5),
                     row_title_gp = gpar(fontsize = 0),
                     gap = unit(0.01, "mm"),
                     use_raster = TRUE,
                     raster_device = "CairoPNG",
                     #heatmap_width = unit(h_width, "cm"),
                     #heatmap_height = unit(h_height, "cm"),
                     raster_quality = 10,
                     show_heatmap_legend = show_legend,
                     heatmap_legend_param = legends_param)
  
  return(ht_FGCs)
}


## Function to draw expression in adult testis cells with specific annotations
## and downasampling the number of cells from each cell type

draw_testis_heatmap <- function(genes, 
                                split_by = "chr_met", 
                                sex = c("F", "M"),
                                ncells_max = NULL,
                                font_size = 6,
                                scale_lims = NULL,
                                clusterRow = FALSE,
                                show_legend = TRUE,
                                h_width = 5,
                                h_height = 10,
                                clust_method = "ward.D"){
  
  testis_adult_sce <- testis_sce[, testis_sce$stage != "somatic"]
  testis_adult_sce$type <- factor(testis_adult_sce$type, levels = germcells)
  testis_adult_sce$stage <- factor(testis_adult_sce$stage, 
                                   levels = c("pre-meiotic", "meiotic", 
                                              "post-meiotic"))
  df_col_testis <- data.frame(type = testis_adult_sce$type,
                              donor = testis_adult_sce$Donor,
                              stage = testis_adult_sce$stage)
  
  df_col_testis$sex <- "M"
  
  rownames(df_col_testis) <- colnames(testis_adult_sce)
  df_col_testis <- df_col_testis[order(df_col_testis$type),]
  
  #Randomly pick ncells_max cells of each cell type
  if (!is.null(ncells_max)) {
    set.seed(123)
    selected_cells <- sample_cells(
      rownames(df_col_testis[df_col_testis$type == 
                               levels(df_col_testis$type)[1],]), ncells_max)
    for (type in levels(df_col_testis$type)[-1]){
      tmp <- sample_cells(
        rownames(df_col_testis[df_col_testis$type == type,]), ncells_max)
      selected_cells <- c(selected_cells, tmp)
    }
    df_col_testis <- df_col_testis[selected_cells,]
  }
  
  column_ha_cell_type = HeatmapAnnotation(
    Testis_cell_type = df_col_testis$type,
    col = list(Testis_cell_type = testis_colors),      
    annotation_legend_param = legends_param,
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_stage = HeatmapAnnotation(
    stage = df_col_testis$stage,
    col = list(stage = stage_colors),
    annotation_legend_param = legends_param,
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_sex = HeatmapAnnotation(
    sex = df_col_testis$sex,
    col = list(sex = sex_colors),
    annotation_legend_param = legends_param,
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_name_gp = gpar(fontsize = 0))
  
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) 
  
  if (split_by == "regulation") {
    gene_type <- gene_type %>% arrange(Regulation)
    rowSplit <- gene_type$Regulation
  }
  if (split_by == "testis_specificity") {
    gene_type <- gene_type %>% arrange(Testis_specificity)
    rowSplit <- gene_type$Testis_specificity
  }
  if (split_by == "chr_met") {
    gene_type <- gene_type %>% arrange(chr_met)
    rowSplit <- gene_type$chr_met
  }
  if (!clusterRow & split_by == "chr_met") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$chr_met
  }
  
  if (split_by == "met_3_groups") {
    gene_type <- gene_type %>% arrange(met_3_groups)
    rowSplit <- gene_type$met_3_groups
  }
  if (!clusterRow & split_by == "met_3_groups") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$met_3_groups
  }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(
    Regulation = Regulation,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(Regulation = Regulation_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(
    chr_met = chr_met,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(chr_met = chr_met_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$chromosome)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              border = FALSE,
                              simple_anno_size = unit(0.15, "cm"), 
                              col = list(Chr = Chr_colors_with_Y),
                              annotation_name_gp = gpar(fontsize = 0))
  
  mat <-  as.matrix(
    assay(testis_adult_sce, "logcounts")[gene_type$gene, rownames(df_col_testis)])
  
  if (is.null(scale_lims)) {
    scale_lims <- c(0, quantile(rowMaxs(mat), 0.8))
  }
  
  heatmap_colors <- colorRamp2(
    seq(scale_lims[1], scale_lims[2], length = 11),                 
    legend_colors)
  
  ht_testis <- Heatmap(mat,
                       name = "logcounts",
                       column_title = "Adult testis",
                       column_split = df_col_testis$type,
                       row_title = NULL,
                       col = heatmap_colors,
                       border = TRUE,
                       border_gp = gpar(lwd = 0.2),
                       row_gap = unit(c(0.5), "mm"),
                       column_gap = unit(c(0.5), "mm"),
                       show_column_dend = FALSE,
                       show_row_dend = FALSE,
                       row_names_gp = gpar(fontsize = font_size),
                       row_names_side = "right",
                       cluster_rows = clusterRow,
                       clustering_method_rows = clust_method,
                       row_dend_reorder = TRUE,
                       cluster_columns = FALSE,
                       show_column_names = FALSE,
                       show_row_names = FALSE,
                       row_split = rowSplit,
                       cluster_row_slices = FALSE,
                       cluster_column_slices = FALSE,
                       top_annotation = c(column_ha_sex, 
                                          column_ha_stage,
                                          column_ha_cell_type),  
                       right_annotation = c(row_ha_regulation, 
                                            row_ha_chr, 
                                            gap = unit(1, "mm")),
                       column_title_gp = gpar(fontsize = 5),
                       row_title_gp = gpar(fontsize = 0),
                       gap = unit(0.01, "mm"),
                       use_raster = TRUE,
                       raster_device = "CairoPNG",
                       raster_quality = 10,
                       #heatmap_width = unit(h_width, "cm"),
                       #heatmap_height = unit(h_height, "cm"),
                       show_heatmap_legend =  show_legend,
                       heatmap_legend_param = legends_param)
  
  return(ht_testis)
}

## Function to draw expression in adult oocytes cells with specific annotations
## and downasampling the number of cells from each cell type

draw_oocytes_heatmap <- function(genes, 
                                 split_by = "chr_met", 
                                 ncells_max = NULL,
                                 font_size = 6,
                                 scale_lims = NULL,
                                 clusterRow = TRUE,
                                 show_legend = TRUE,
                                 h_width = 5,
                                 h_height = 10,
                                 clust_method = "ward.D"){
  
  adult_oocytes_sce <- oocytes_sce[, oocytes_sce$type %in% c("Growing oocytes", 
                                                             #"Fully grown oocytes",  
                                                             #"Metaphase I", 
                                                             "Metaphase II")]
  
  df_col_oocytes <- data.frame(type = adult_oocytes_sce$type,
                               sex = adult_oocytes_sce$sex,
                               stage = adult_oocytes_sce$stage)
  
  rownames(df_col_oocytes) <- colnames(adult_oocytes_sce)
  df_col_oocytes <- df_col_oocytes[order(df_col_oocytes$type),]
  
  #Randomly pick ncells_max cells of each cell type
  if (!is.null(ncells_max)) {
    set.seed(123)
    selected_cells <- sample_cells(
      rownames(df_col_oocytes[df_col_oocytes$type == 
                                levels(df_col_oocytes$type)[1],]), ncells_max)
    for (type in levels(df_col_oocytes$type)[-1]){
      tmp <- sample_cells(
        rownames(df_col_oocytes[df_col_oocytes$type == type,]), ncells_max)
      selected_cells <- c(selected_cells, tmp)
    }
    df_col_oocytes <- df_col_oocytes[selected_cells,]
  }
  
  column_ha_cell_type = HeatmapAnnotation(
    Oocytes_cell_type = df_col_oocytes$type,
    col = list(Oocytes_cell_type = oocytes_colors),     
    border = TRUE,
    simple_anno_size = unit(0.15, "cm"),
    annotation_name_gp = gpar(fontsize = 0),
    annotation_legend_param = legends_param)
  
  column_ha_stage = HeatmapAnnotation(
    stage = df_col_oocytes$stage,
    col = list(stage = stage_colors),
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_sex = HeatmapAnnotation(
    sex = df_col_oocytes$sex,
    col = list(sex = sex_colors),
    annotation_legend_param = legends_param,
    border = TRUE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_name_gp = gpar(fontsize = 0))
  
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) 
  
  if (split_by == "regulation") {
    gene_type <- gene_type %>% arrange(Regulation)
    rowSplit <- gene_type$Regulation
  }
  if (split_by == "testis_specificity") {
    gene_type <- gene_type %>% arrange(Testis_specificity)
    rowSplit <- gene_type$Testis_specificity
  }
  if (split_by == "chr_met") {
    gene_type <- gene_type %>% arrange(chr_met)
    rowSplit <- gene_type$chr_met
  }
  if (!clusterRow & split_by == "chr_met") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$chr_met
  }
  if (split_by == "met_3_groups") {
    gene_type <- gene_type %>% arrange(met_3_groups)
    rowSplit <- gene_type$met_3_groups
  }
  if (!clusterRow & split_by == "met_3_groups") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$met_3_groups
  }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(
    Regulation = Regulation,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(Regulation = Regulation_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(
    chr_met = chr_met,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(chr_met = chr_met_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$chromosome)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              border = FALSE,
                              simple_anno_size = unit(0.15, "cm"), 
                              col = list(Chr = Chr_colors_with_Y),
                              annotation_name_gp = gpar(fontsize = 0))
  
  mat <-  as.matrix(
    assay(adult_oocytes_sce, "logcounts")[gene_type$gene, rownames(df_col_oocytes)])
  
  if (is.null(scale_lims)) {
    scale_lims <- c(0, quantile(rowMaxs(mat), 0.8))
  }
  
  heatmap_colors <- colorRamp2(
    seq(scale_lims[1], scale_lims[2], length = 11),                 
    legend_colors)
  
  ht_oocytes <- Heatmap(mat,
                        name = "oocytes logcounts",
                        column_title = "Adult oocytes",
                        column_split = df_col_oocytes$type,
                        col = heatmap_colors,
                        border = TRUE,
                        border_gp = gpar(lwd = 0.2),
                        row_gap = unit(c(0.5), "mm"),
                        column_gap = unit(c(0.5), "mm"),
                        show_column_dend = FALSE,
                        show_row_dend = FALSE,
                        row_names_gp = gpar(fontsize = font_size),
                        row_names_side = "left",
                        cluster_rows = clusterRow,
                        clustering_method_rows = clust_method,
                        row_dend_reorder = TRUE,
                        cluster_columns = FALSE,
                        show_column_names = FALSE,
                        show_row_names = TRUE,
                        row_split = rowSplit,
                        cluster_row_slices = FALSE,
                        cluster_column_slices = FALSE,
                        top_annotation = c(column_ha_sex, 
                                           column_ha_stage,
                                           column_ha_cell_type),  
                        # right_annotation = c(row_ha_regulation, 
                        #                      row_ha_chr, 
                        #                      gap = unit(1, "mm")),
                        column_title_gp = gpar(fontsize = 5),
                        row_title_gp = gpar(fontsize = 0),
                        gap = unit(0.01, "mm"),
                        use_raster = TRUE,
                        raster_device = "CairoPNG",
                        raster_quality = 10,
                        #heatmap_width = unit(h_width, "cm"),
                        #heatmap_height = unit(h_height, "cm"),
                        show_heatmap_legend = show_legend,
                        heatmap_legend_param = legends_param)
  
  return(ht_oocytes)
}
```

```{r draw_healthy_cells_exprs}
# Manual adjustments to visualise
X_met_new_order <- 
  c("SPANXB1", "SPANXD", "SPANXC",  "CXorf51B", "DDX53", 
    "TGIF2LX", "FTHL17","VCX3B", "VCX", "VCX3A", "VCX2", "FMR1NB",  "PAGE2", 
    "PAGE5", "MAGEB6", "MAGEC2","GAGE2A","PAGE1",  "CTAG2",  "MAGEA4", 
    "PASD1","MAGEB2","CT55","MAGEA10",  "MAGEA3",  "MAGEA6","GAGE1", "MAGEA1", 
    "CT45A10", "MAGEA11", "TEX13C", "SSX1", "MAGEA9B","MAGEA9", "MAGEA8", 
    "MAGEB1", "MAGEA12", "MAGEC1", "MAGEB16","GAGE12E",  "XAGE5", "GAGE13", 
    "SSX5",  "SSX2B", "XAGE1B",  "DCAF8L2", "SSX2", "SAGE1", 
    "CXorf49", "CTAG1A", "NXF2",  "CTAG1B", "XAGE1A", 
    "CXorf49B","PPP4R3C", "CT47A12",
    "CT47A11", "CT45A1","CT45A5","CT45A7", "MAGEA4-AS1", "MAGEA2",
      "CENPVL2", "CENPVL1", "CT47A1",  "CT45A3",  "CT45A6", "CT45A8", "CT45A9")

not_X_met_new_order <- 
  c( "FLJ36000", "TRIML2", "LINC01413", "NLRP4", "SUN3",
    "LINC02475",  "TUBA3C", "COX7B2", 
     "BRDT", "DSCR8",  "RBM46", "RNF17", "DCAF4L2",
     "POTEC","CTCFL", "PRDM9",  "LINC00470", 
    "TMEM132D-AS1", "DPPA2", 
     "LINC01518", "LINC00200", "RFPL4B", "POTEF", 
      "GFY", "SLC7A11-AS1",
    "KAZN-AS1", "LINC02377",    "LINC02241",   "SPATA31C2",   
    "FAM230C", "GOLGA6L7", "TEX48", "LINC02663", "LINC02864",
    "RBMY1E", "VCY", "VCY1B", "RBMY1A1","RBMY1B"   ) 


not_met_new_order <- 
  c("CHCT1", "RAD21L1", "CSTL1", "WDR64", "CBY2", "GARIN2","LINC01193", 
    "FAM217A", "LIN28A", "LINC01249", 
    "CDRT15","C1orf94", "LINC01206", "SYCP1", "MEIOB", "MAJIN",
    "PRDM7",  "LINC01098", "ESX1", "DMRT1", "LINC01511", "ZNF729", 
    "NFIA-AS1", "THORLNC", "LINC02228", "SEC61G-DT", "ZNF722",  "ZNF723",
    "LRRC52-AS1",
    
     "LINC03004", 
    "GTF2I-AS1", "LINC02750",  "TSPY4", "TSPY1", "TSPY9", "TSPY3", "TSPY2") 


my_genes <- c(X_met_new_order, not_X_met_new_order, not_met_new_order)


fs <- 2
scale_limits <- c(0, 4)
ncells <- 100
ht_testis <- draw_testis_heatmap(my_genes,
                                 ncells_max = ncells,
                                 font_size = fs,
                                 scale_lims = scale_limits,
                                 clusterRow  = FALSE,
                                 split_by = "met_3_groups",
                                 show_legend = TRUE,
                                 clust_method = "ward.D2")

female_FGC_ht <- draw_FGC_heatmap(my_genes, 
                                  ncells_max = ncells,
                                  sex = c("F"),
                                  clusterRow = FALSE, 
                                  font_size = fs,
                                  scale_lims = scale_limits, 
                                  show_legend = FALSE,
                                  show_right_annot = FALSE)

male_FGC_ht <- draw_FGC_heatmap(my_genes, 
                                ncells_max = ncells,
                                sex = c("M"),
                                clusterRow = FALSE, 
                                font_size = fs,
                                scale_lims = scale_limits, 
                                show_legend = FALSE,
                                show_right_annot = FALSE)


ht_oocytes <- draw_oocytes_heatmap(my_genes, 
                                   ncells_max = ncells,
                                   font_size = fs,
                                   scale_lims = scale_limits, 
                                   clusterRow = FALSE,
                                   show_legend = FALSE,
                                   clust_method = "centroid")

ht_list <- female_FGC_ht + ht_oocytes + male_FGC_ht + ht_testis
draw(ht_list, main_heatmap = "logcounts", merge_legend = TRUE, 
     ht_gap = unit(0.2, "cm"))


```

```{r, echo = FALSE, eval = FALSE}
pdf(file = paste0("figs/CT_in_GermCells.pdf"), width = 6, height = 4)
draw(ht_list, main_heatmap = "logcounts", 
     merge_legend = TRUE, 
     ht_gap = unit(0.2, "cm"), 
     legend_gap = unit(0.15, "cm"),
     legend_grid_height = unit(0.08, "cm"),
     legend_grid_width = unit(0.1, "cm"))
dev.off()
```

## Numbers

```{r}
CT_genes_not_Y <- CT_genes %>% 
  filter(chr != "Y") %>% 
  filter(!external_gene_name %in% never) %>% 
  pull(external_gene_name)

dds_FCE <- scuttle::summarizeAssayByGroup(FGC_sce, 
                            ids = colData(FGC_sce)[, "type"],
                            statistics = "prop.detected")
dds_oocytes <- scuttle::summarizeAssayByGroup(oocytes_sce, 
                            ids = colData(oocytes_sce)[, "type"],
                            statistics = "prop.detected")
testis_Germ_cells <- testis_sce[, testis_sce$type %in% germcells]
dds_testis <- scuttle::summarizeAssayByGroup(testis_Germ_cells, 
                            ids = colData(testis_Germ_cells)[, "type"],
                            statistics = "prop.detected")


Male_cells <- cbind(assay(dds_testis[CT_genes_not_Y, ]) > 0.1, 
      assay(dds_FCE[CT_genes_not_Y, c("M_PGC", "M_GC", 
                                      "M_pre_spermatogonia")]) > 0.1)

Female_cells <- cbind(assay(dds_FCE[CT_genes_not_Y, 
                                    c("F_PGC", "F_GC", 
                                      "F_oogonia", "F_oocyte")]) > 0.1, 
      assay(dds_oocytes[CT_genes_not_Y, c("Growing oocytes", 
                                          "Metaphase II")]) > 0.1)

prop_CT_genes_by_gender <- tibble(gene = CT_genes_not_Y, 
                                  in_female = rowSums(Female_cells) > 0,
       in_male = rowSums(Male_cells) > 0) %>% 
  summarize(Female = sum(in_female)/length(CT_genes_not_Y),
            Male = sum(in_male)/length(CT_genes_not_Y)) %>% 
  mutate(gene = "CT gene") %>% 
  pivot_longer(names_to = "Germ cell", 
               values_to = "Proportion of CT genes detected in germ cells",
               -gene) 

A <- prop_CT_genes_by_gender %>% 
  ggplot(aes(x = `Germ cell`, 
             y = `Proportion of CT genes detected in germ cells`, 
             fill = `Germ cell`)) +
  geom_col() +
  scale_fill_manual(values = c("Female" = "deeppink", "Male" = "steelblue")) +
  theme_bw() +
  theme(aspect.ratio = 1.8) +
  xlab('') +
  theme(legend.position = "none")

prop_CT_genes_by_gender_by_group <- 
  tibble(gene = CT_genes_not_Y, in_female = rowSums(Female_cells) > 0,
in_male = rowSums(Male_cells) > 0) %>% 
  left_join(CT_genes %>% 
              dplyr::rename(gene = external_gene_name) %>% 
              dplyr::select(gene, X_linked, regulated_by_methylation)) %>% 
  mutate(group = case_when(X_linked & regulated_by_methylation ~ "met\nX-linked",
                           !X_linked & regulated_by_methylation ~ "met\nAutosome",
                           !regulated_by_methylation ~ "Not\nmet")) %>% 
  mutate(group = factor(group, levels = c("met\nX-linked", "met\nAutosome", 
                                          "Not\nmet"))) %>% 
  group_by(group) %>% 
  summarize(Female = sum(in_female), Male = sum(in_male)) %>% 
  pivot_longer(names_to = "Germ cell", 
               values_to = "Number of genes detected", -group) %>% 
  filter(!is.na(group))

B <-  prop_CT_genes_by_gender_by_group %>%  
  ggplot(aes(x = group, y = `Number of genes detected`, fill = `Germ cell`)) +
  geom_col() +
   scale_fill_manual(values = c("Female" = "deeppink", "Male" = "steelblue")) +
  theme_bw() +
  theme(aspect.ratio = 1.8) +
  xlab('') 

x <- tibble(gene = CT_genes_not_Y, in_female = rowSums(Female_cells) > 0,
in_male = rowSums(Male_cells) > 0) %>% 
  left_join(CT_genes %>% dplyr::rename(gene = external_gene_name) %>% 
              dplyr::select(gene, X_linked, regulated_by_methylation)) %>% 
  mutate(group = case_when(X_linked & regulated_by_methylation ~ "met\nX-linked",
                           !X_linked & regulated_by_methylation ~ "met\nAutosome",
                           !regulated_by_methylation ~ "Not\nmet")) %>% 
  mutate(group = factor(group, 
                        levels = c("met\nX-linked", "met\nAutosome", 
                                   "Not\nmet"))) %>% 
  group_by(group) %>% 
  summarize(Female = sum(in_female), Male = sum(in_male)) %>% 
  filter(!is.na(group))
x
chisq.test(x[,-1])


fig <- cowplot::plot_grid(A + theme(legend.position = "none"), 
                          B + theme(legend.position = "none"),
           rel_heights = c(1, 1))

legend <- cowplot::get_legend(B)
cowplot::plot_grid(fig, legend, align = "v", rel_widths = c(1, 0.2))
```

```{r, echo = FALSE, eval = FALSE}
pdf(file = paste0("figs/Prop_Female_Male.pdf"), width = 6.8, height = 4)
cowplot::plot_grid(fig, legend,
                rel_widths = c(2, 1))
dev.off()
```

# CT methylation in fetal germ cells

```{r}
scWGBS_FGC_SE$cellType <- factor(scWGBS_FGC_SE$cellType, 
                          levels = c("F_FGC", "F_meiotic_FGC","F_oogenesis_FGC",
                                     "M_mitotic_FGC", "M_mitotic_arrested_FGC",
                                     "F_soma","M_soma"))

# Changed cellType names for coherence with expression heatmap
scWGBS_FGC_SE$cellType <- as.vector(scWGBS_FGC_SE$cellType)
scWGBS_FGC_SE$cellType[scWGBS_FGC_SE$cellType == "F_FGC"] <- "F_GC"
scWGBS_FGC_SE$cellType[scWGBS_FGC_SE$cellType == "F_meiotic_FGC"] <- "F_oogonia"
scWGBS_FGC_SE$cellType[scWGBS_FGC_SE$cellType == "F_oogenesis_FGC"] <- "F_oocyte"
scWGBS_FGC_SE$cellType[scWGBS_FGC_SE$cellType == "M_mitotic_FGC"] <- "M_GC"
scWGBS_FGC_SE$cellType[scWGBS_FGC_SE$cellType == "M_mitotic_arrested_FGC"] <- "M_pre_spermatogonia"


scWGBS_FGC_SE$cellType <- factor(scWGBS_FGC_SE$cellType, 
                          levels = c("F_GC", "F_oogonia","F_oocyte",
                                     "M_GC", "M_pre_spermatogonia",
                                     "F_soma","M_soma"))
```

## Mean meth values heatmap

```{r, warning = FALSE, message = FALSE}
nt_up <- 1000
nt_down <- 500

CT_promoter_regions <- as_tibble(CT_promoter_regions_hg19_GR) %>% 
  dplyr::select(external_gene_name, chr, TSS_liftover, strand) %>%
  mutate(start = case_when(strand == "+" ~ TSS_liftover - nt_up,
                           strand == "-" ~ TSS_liftover - nt_down)) %>%
  mutate(stop = case_when(strand == "+" ~ TSS_liftover + nt_down,
                          strand == "-" ~ TSS_liftover + nt_up)) %>%
  mutate(chromosome_name = paste0("chr", chr)) 

CT_prom_GR <- GenomicRanges::makeGRangesFromDataFrame(CT_promoter_regions,
                                                      keep.extra.columns = TRUE,
                                                      seqnames.field = "chromosome_name",
                                                      start.field = "start",
                                                      end.field = "stop")


tested_gene_GR <- subsetByOverlaps(scWGBS_FGC_SE, CT_promoter_regions_hg19_GR[1])
mat <- as.matrix(assay(tested_gene_GR))
met_means <- enframe(colMeans(mat, na.rm = TRUE))
names(met_means) <- c("sample", CT_promoter_regions_hg19_GR[1]$external_gene_name)


for (i in 2:length(CT_promoter_regions_hg19_GR)){
    tested_gene_GR <- subsetByOverlaps(scWGBS_FGC_SE, CT_promoter_regions_hg19_GR[i])

      mat <- as.matrix(assay(tested_gene_GR))
      
      tmp <- enframe(colMeans(mat, na.rm = TRUE))
      names(tmp) <- c("sample", CT_promoter_regions_hg19_GR[i]$external_gene_name)
      met_means <- left_join(met_means, tmp)
  }
  
met_means <- met_means %>% pivot_longer(names_to = "gene", values_to = "met", -sample) %>% 
      pivot_wider(names_from = sample, values_from = met)
```


Methylation in promoter regions defined as `r, nt_up` nt upstream TSS 
and `r, nt_down` nt downstream TSS 

```{r}
# add NA when missing CT genes 

missing_CT <- CT_genes$external_gene_name[!CT_genes$external_gene_name %in% met_means$gene]
met_means_missing_CT <- met_means[1:length(missing_CT),]
met_means_missing_CT$gene <- missing_CT
met_means_missing_CT[, 2:ncol(met_means_missing_CT)] <- NA
met_means <- rbind(met_means, met_means_missing_CT)

plot_mean_methylation_heatmap <- function(genes, 
                                          clusterRows = F,
                                          fs = 8){
  
  df_col <- colData(scWGBS_FGC_SE)
  df_col <- df_col[order(df_col$cellType),] 
  
  column_ha_time = HeatmapAnnotation(
    time = df_col$time,
    col = list(time = Fetal_time_colors),
    border = T, simple_anno_size = unit(0.4, "cm"))
  
  column_ha_sex = HeatmapAnnotation(
    sex = df_col$sex,
    col = list(sex = sex_colors),
    border = T, simple_anno_size = unit(0.4, "cm"))
  
  column_ha_cellType = HeatmapAnnotation(
    cellType = df_col$cellType,
    col = list(cellType = Fetal_cellType_colors),
    border = T, simple_anno_size = unit(0.4, "cm"))
  

  genes <- genes[genes %in% met_means$gene]
  mat <- as.matrix(met_means[,-1])
  rownames(mat)<- met_means$gene
  mat <- mat[genes, df_col$sample]

  # Rm cells with no methylation values
  # mat <- mat[rowSums(!is.na(mat)) > 0, ,drop = F]
  # df_col <- df_col[rownames(mat),]
    

  h <- Heatmap(mat,
               column_title = "Mean methylation in promoters",
               name = 'Met',
               column_split = df_col$group,
               row_gap = unit(c(2), "mm"),
               row_title_gp = gpar(fontsize = 8, fontface = "bold"),
               column_title_gp = gpar(fontsize = 8),
               row_title_rot = 0,
               col = colorRamp2(c(1:100),colorRampPalette(
                 c("moccasin","dodgerblue4"))(100)),
               na_col = "gray80",
               clustering_method_rows = "ward.D",
               cluster_rows = clusterRows,
               cluster_columns = FALSE,
               show_row_names = TRUE,
               show_column_names = TRUE,
               show_heatmap_legend = TRUE,
               show_row_dend = FALSE,
               top_annotation = c(column_ha_cellType,
                                  column_ha_time,
                                  column_ha_sex,
                                  gap = unit(0.1, "mm")),
               # right_annotation = c(row_ha_week,  row_ha_sex,
               #                      row_ha_type,
               #                      gap = unit(1, "mm")),
               row_names_gp = gpar(fontsize = fs),
               column_names_gp = gpar(fontsize = 8),
               column_names_side = "bottom",
              # width = unit(0.2, "npc"),
               row_names_side = "left")
  
  print(h) 
}
```

  

```{r}
X_met_new_order <- 
  c("PASD1","TEX13C",  "SSX1", 
    "VCX3B", "VCX3A", "VCX2", "FMR1NB", "MAGEB6", "MAGEC2","GAGE2A","PAGE1",
    "MAGEA1", 
    "CT45A10", "CT45A1","SSX5",  "SAGE1", "PPP4R3C","MAGEA12", "MAGEC1", "MAGEB16","MAGEA3",
    "MAGEA6","GAGE1", "MAGEA9B","MAGEA9", "GAGE12E", "GAGE13", "MAGEA2", 
    "CXorf49", "VCX", 
    "PAGE2", 
    "PAGE5","CTAG2","MAGEA8", 
    "MAGEB1", 
    "MAGEA4", 
    "TGIF2LX", "FTHL17",
    "MAGEB2","CT55","MAGEA10",    "MAGEA11",  
    "XAGE5", 
    "SPANXD",
    "DCAF8L2", "SPANXC", 
    "MAGEA4-AS1",  "DDX53",
    "CENPVL1", "SPANXB1","CXorf51B", "SSX2B",
    "XAGE1B",  "SSX2", "CTAG1A", "NXF2",  "CTAG1B", "XAGE1A",
    "CXorf49B", "CT47A12",
    "CT47A11","CT45A5","CT45A7", "CENPVL2","CT47A1",
    "CT45A3",  "CT45A6",  "CT45A8",  "CT45A9" )

not_X_met_new_order <- 
  c( "LINC01518", "POTEF", "KAZN-AS1","GOLGA6L7", "TUBA3C", "COX7B2",  "RNF17", 
     "DCAF4L2","DPPA2", "LINC02663",
     "LINC00200", "RFPL4B", 
     "LINC01413", "NLRP4", "SUN3",
     "LINC02475",  
     "BRDT", "DSCR8",  "RBM46", 
     "POTEC","CTCFL", "PRDM9",  "LINC00470", 
     "TMEM132D-AS1", 
     "SLC7A11-AS1",
     "LINC02377",    "LINC02241",   "SPATA31C2",   
     "TEX48",  "TRIML2", "LINC02864", "GFY",
"FLJ36000", "FAM230C", "RBMY1E", "VCY", "VCY1B", "RBMY1A1","RBMY1B"   ) 


not_met_new_order <- 
  c("LINC01098",  "LINC01511", "ZNF722", "LRRC52-AS1", 
    "CSTL1", "WDR64", "CBY2", "GARIN2","LINC01193", 
    "FAM217A",  "LINC01249", 
    "CDRT15", "LINC01206", "SYCP1", "MEIOB", "MAJIN",
    "PRDM7",  
    "NFIA-AS1",  "LINC02228",  "ZNF723",
    "LINC03004", 
    "GTF2I-AS1", "LINC02750",  "CHCT1", "RAD21L1", "LIN28A","C1orf94", 
    "ESX1", "DMRT1", "ZNF729", "THORLNC","SEC61G-DT", "TSPY4", "TSPY1", 
    "TSPY9", "TSPY3", "TSPY2") 


my_genes <- c(X_met_new_order, not_X_met_new_order, not_met_new_order)
```


```{r}
# slightly modified the functions to create the final heatmap
draw_FGC_heatmap <- function(genes, 
                             split_by = "chr_met", 
                             ncells_max = NULL,
                             cell_type = NULL,
                             scale_lims = NULL,
                             sex = c("F", "M"),
                             clusterRow = FALSE, 
                             font_size = 6,
                             show_legend = TRUE,
                             h_width = 5,
                             h_height = 10,
                             show_right_annot = TRUE){
  
  fetal_sce <- FGC_sce
  
  if (!is.null(cell_type)) fetal_sce <- fetal_sce[,FGC_sce$type %in% cell_type]
  
  # Set the colors used in the final heatmap based on all samples (F + M), 
  # even when representing only F or M samples on the heatmap
  # => same scale of colors
  mat <-  as.matrix(assay(fetal_sce, "logcounts")[genes, ])
  
  if (is.null(scale_lims)) {
    scale_lims <- c(0, quantile(rowMaxs(mat), 0.8))
  }
  
  heatmap_colors <- colorRamp2(
    seq(scale_lims[1], scale_lims[2], length = 11),                 
    legend_colors)
  
  fetal_sce <- fetal_sce[, fetal_sce$sex %in% sex]
  
  df_col_FCS <- data.frame(development_stage = fetal_sce$development_stage,
                           sex = fetal_sce$sex,
                           stage = fetal_sce$stage,
                           donor_id = fetal_sce$donor_id,
                           fetal_cell_type = fetal_sce$cell_type,
                           type = fetal_sce$type)

  
  rownames(df_col_FCS) <- colnames(fetal_sce)
  df_col_FCS <- df_col_FCS[order(df_col_FCS$type),]
  
  #Randomly pick ncells_max cells of each cell type
  if (!is.null(ncells_max)) {
    set.seed(123)
    selected_cells <- sample_cells(
      rownames(df_col_FCS[df_col_FCS$type == levels(df_col_FCS$type)[1],]), 
      ncells_max)
    for (type in levels(df_col_FCS$type)[-1]){
      tmp <- sample_cells(
        rownames(df_col_FCS[df_col_FCS$type == type,]), ncells_max)
      selected_cells <- c(selected_cells, tmp)
    }
    df_col_FCS <- df_col_FCS[selected_cells,]
  }
  
  column_ha_type = HeatmapAnnotation(
    Fetal_cell_type = df_col_FCS$type,
    col = list(Fetal_cell_type = Fetal_cell_colors),   
    border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0),
    show_legend = FALSE)
  
  column_ha_development_stage = HeatmapAnnotation(
    development_stage = df_col_FCS$development_stage,
    border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_sex = HeatmapAnnotation(
    sex = df_col_FCS$sex,
    col = list(sex = sex_colors),
    border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_donor_id = HeatmapAnnotation(
    donor_id = df_col_FCS$donor_id,
    border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_stage = HeatmapAnnotation( 
    stage = df_col_FCS$stage,
    col = list(stage = stage_colors),
    border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) 
  
  if (split_by == "regulation") {
    gene_type <- gene_type %>% arrange(Regulation)
    rowSplit <- gene_type$Regulation
  }
  if (split_by == "testis_specificity") {
    gene_type <- gene_type %>% arrange(Testis_specificity)
    rowSplit <- gene_type$Testis_specificity
  }
  if (split_by == "chr_met") {
    gene_type <- gene_type %>% arrange(chr_met)
    rowSplit <- gene_type$chr_met
  }
  if (!clusterRow & split_by == "chr_met") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$chr_met
  }
  if (split_by == "met_3_groups") {
    gene_type <- gene_type %>% arrange(met_3_groups)
    rowSplit <- gene_type$met_3_groups
  }
   if (!clusterRow & split_by == "met_3_groups") {
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$met_3_groups
   }
  
  Regulation <- as.matrix(gene_type$Regulation)
  rownames(Regulation) <- gene_type$gene
  row_ha_regulation <- rowAnnotation(
    Regulation = Regulation,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(Regulation = Regulation_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  chr_met <- as.matrix(gene_type$chr_met)
  rownames(chr_met) <- gene_type$gene
  row_ha_chr_met <- rowAnnotation(
    chr_met = chr_met,
    annotation_legend_param = legends_param,
    border = FALSE,
    simple_anno_size = unit(0.15, "cm"), 
    col = list(chr_met = chr_met_colors),
    annotation_name_gp = gpar(fontsize = 0))
  
  Chr <- as.matrix(gene_type$chromosome)
  rownames(Chr) <- gene_type$gene
  row_ha_chr <- rowAnnotation(Chr = Chr,
                              annotation_legend_param = legends_param,
                              border = FALSE,
                              simple_anno_size = unit(0.15, "cm"), 
                              col = list(Chr = Chr_colors_with_Y),
                              annotation_name_gp = gpar(fontsize = 0))
  
  mat <- as.matrix(
    assay(fetal_sce, "logcounts")[gene_type$gene, rownames(df_col_FCS)])
  
  rAnnot <- c(row_ha_regulation, row_ha_chr, gap = unit(1, "mm"))
  
  if (!show_right_annot) rAnnot <- NULL
  
  title <- "Expression"
  
  if (is.null(scale_lims)) {
    scale_lims <- c(0, quantile(rowMaxs(mat), 0.8))
  }
  
  heatmap_colors <- colorRamp2(
    seq(scale_lims[1], scale_lims[2], length = 11),                 
    legend_colors)

  ht_FGCs <- Heatmap(mat,
                     name = "Logcounts",
                     column_title = title,
                     column_split = df_col_FCS$sex,
                     cluster_row_slices = FALSE,
                     cluster_column_slices = FALSE,
                     row_title_gp = gpar(fontsize = 3),
                     row_title_rot = 90,
                     col = heatmap_colors,
                     border = TRUE,
                     border_gp = gpar(lwd = 0.2),
                     row_gap = unit(c(0.4), "mm"),
                     column_gap = unit(c(0.4), "mm"),
                     show_column_dend = FALSE,
                     show_row_dend = FALSE,
                     row_names_gp = gpar(fontsize = font_size),
                     row_names_side = "left",
                     cluster_rows = clusterRow,
                     cluster_columns = FALSE,
                     show_column_names = FALSE,
                     show_row_names = TRUE,
                     row_split = rowSplit,
                     top_annotation = c(column_ha_sex, 
                                        #column_ha_stage,
                                        #column_ha_group, 
                                        #column_ha_donor_id,
                                        column_ha_type,
                                        gap = unit(0.1, "mm")), 
                     right_annotation = rAnnot,
                     column_title_gp = gpar(fontsize = 4),
                     gap = unit(0.01, "mm"),
                     use_raster = TRUE,
                     raster_device = "CairoPNG",
                     heatmap_width = unit(h_width, "cm"),
                     heatmap_height = unit(h_height, "cm"),
                     raster_quality = 10,
                     show_heatmap_legend = show_legend,
                     heatmap_legend_param = legends_param)
  
  return(ht_FGCs)
}
```



```{r}
# Modify the function to keep the initial order of genes, 
# grouping by the met_3groups column
plot_mean_methylation_heatmap_met_3_groups <- function(genes,
                                                       fs = 8,
                                                       h_width = 3,
                                                       h_height = 8){
  
  df_col <- colData(scWGBS_FGC_SE)
  df_col <- df_col[order(df_col$cellType),] 
  
  column_ha_time = HeatmapAnnotation(
    time = df_col$time,
    col = list(time = Fetal_time_colors),   
    border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_sex = HeatmapAnnotation(
    sex = df_col$sex,
    col = list(sex = sex_colors),   
   border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  column_ha_cellType = HeatmapAnnotation(
    cellType = df_col$cellType,
    col = list(cellType = Fetal_cellType_colors),
      border = FALSE, 
    simple_anno_size = unit(0.15, "cm"), 
    annotation_legend_param = legends_param,
    annotation_name_gp = gpar(fontsize = 0))
  
  # add NA values to missing genes
  missing_genes <- CT_genes_types %>%
    #filter(gene %in% genes) %>%
    filter(!gene %in% met_means$gene)
  mat <- as.matrix(met_means[,-1])
  rownames(mat)<- met_means$gene
  mat_missing_values <- matrix(NA, nrow(missing_genes), ncol(mat))
  rownames(mat_missing_values) <- missing_genes$gene
  colnames(mat_missing_values) <- colnames(mat)
  mat <- rbind(mat, mat_missing_values)
  
  #genes <- genes[genes %in% met_means$gene]
  mat <- mat[genes, df_col$sample]
  
  gene_type <- CT_genes_types %>% 
    filter(gene %in% genes) 

  # split_by "met_3_groups"
    gene_type$gene <- factor(gene_type$gene, levels = genes)
    gene_type <- gene_type  %>%
      arrange(gene)
    gene_type$gene <- as.vector(gene_type$gene)
    rowSplit <- gene_type$met_3_groups

  mat <- mat[genes, df_col$sample]

  # Rm cells with no methylation values
  # mat <- mat[rowSums(!is.na(mat)) > 0, ,drop = F]
  # df_col <- df_col[rownames(mat),]
    

  h <- Heatmap(mat,
               column_title = "Mean methylation",
               name = 'Met',
               column_split = df_col$sex_type,
               row_split = rowSplit,
               cluster_row_slices = FALSE,
               border_gp = gpar(lwd = 0.2),
               row_gap = unit(c(0.4), "mm"),
               column_gap = unit(c(0.4), "mm"),
               row_title_gp = gpar(fontsize = 5),
               row_title_rot = 90,
               column_title_gp = gpar(fontsize = 4),
               col = colorRamp2(c(1:100),colorRampPalette(
                 c("moccasin","dodgerblue4"))(100)),
               na_col = "gray80",
               clustering_method_rows = "ward.D",
               cluster_rows = FALSE,
               cluster_columns = FALSE,
               show_row_names = TRUE,
               show_column_names = FALSE,
               show_heatmap_legend = TRUE,
               show_row_dend = FALSE,
               top_annotation = c(#column_ha_time, 
                                  column_ha_sex,
                                  column_ha_cellType,
                                  gap = unit(0.1, "mm")), 
               row_names_gp = gpar(fontsize = fs),
               column_names_gp = gpar(fontsize = 3),
               column_names_side = "top",
               gap = unit(0.01, "mm"),
               use_raster = TRUE,
               raster_device = "CairoPNG",
               heatmap_width = unit(h_width, "cm"),
               heatmap_height = unit(h_height, "cm"),
               raster_quality = 20,
               row_names_side = "left",
               heatmap_legend_param = legends_param)
  
  print(h) 
}
```

```{r}
legends_param <- list(
  labels_gp = gpar(col = "black", fontsize = 2),
  title_gp = gpar(col = "black", fontsize = 3),
  simple_anno_size = unit(0.15, "cm"),
  row_names_gp = gpar(fontsize = 1),
  annotation_name_side = "left",
  border = FALSE,
  border_gp = gpar(lwd = 0.01),
  grid_width = unit(0.1, "cm"),
  grid_height = unit(0.2, "cm"),
  legend_height = unit(0.1, "cm"),
  annotation_name_gp = gpar(fontsize = 0))

FGC_ht <- draw_FGC_heatmap(my_genes, 
                           ncells_max = 100,
                           split_by = "met_3_groups",
                           cell_type = c("F_GC", "F_oogonia", "F_oocyte", 
                                         "M_GC", "M_pre_spermatogonia"),
                           #sex = c("F"),
                           clusterRow = FALSE, 
                           font_size = 1,
                           scale_lims = scale_limits, 
                           show_legend = TRUE,
                           h_width = 2,
                           h_height = 6,
                           show_right_annot = FALSE)


fig <- plot_mean_methylation_heatmap_met_3_groups(my_genes, 
                                                  h_width = 3,
                                                  h_height = 6,
                                                  fs  = 2)


ht_list <- FGC_ht + fig 

draw(ht_list,  
     merge_legend = TRUE)

```
```{r, echo = FALSE, eval = FALSE}
pdf(file = paste0("figs/Exp_methylation_in_FGC.pdf"), width = 5, height = 3)
draw(ht_list,
     merge_legend = TRUE)
dev.off()
```


## Gene by gene

Change to hg19 coordinates + heatmap

```{r}
promoter_methylation_by_gene <- function(gene, 
                                         nt_up = 1000, 
                                         nt_down = 500,
                                         cellType = NULL,
                                         sex = NULL,
                                         h_width = 5,
                                         h_height = 5,
                                         fs = 8,
                                         col_fs = 4,
                                         title = NULL){
  
  
  x <- CT_promoter_regions_hg19_GR[
    CT_promoter_regions_hg19_GR$external_gene_name == gene, ]
  tested_gene_GR <- subsetByOverlaps(scWGBS_FGC_SE, x)
  CpGpos <- tested_gene_GR@rowRanges@ranges@start
  TSS <- x$TSS_liftover
  if(as.vector(x@strand) == '+'){
    CpGpositions <- CpGpos - x$TSS_liftover
  }
  
  if(as.vector(x@strand) == '-'){
    CpGpositions <- x$TSS_liftover  - CpGpos 
  }
  df_col <- colData(tested_gene_GR)
  
  if (!is.null(cellType)) df_col <- df_col[df_col$cellType %in% cellType, ]
  if (!is.null(sex)) df_col <- df_col[df_col$sex %in% sex, ]
  
  df_col <- df_col[order(df_col$cellType),] 
  levels(df_col$cellType) <- gsub("^\\w_", '', levels(df_col$cellType))
  
  y <- as_tibble(assay(tested_gene_GR)[, df_col$sample]) %>% 
    mutate(CpG = CpGpositions) %>% dplyr::select(CpG, everything()) %>% 
    filter(CpG > -nt_up & CpG < nt_down) %>% 
    arrange(CpG) %>% 
    pivot_longer(names_to = "sample", values_to = "met", -CpG) %>% 
    pivot_wider(names_from = CpG, values_from = met)
  
  mat <- as.matrix(y[, -1])
  rownames(mat) <- y$sample
  
  # Rm cells with no methylation values
  mat <- mat[rowSums(!is.na(mat)) > 5, ,drop = F]
  df_col <- df_col[rownames(mat),]
  
  sex <- as.matrix(df_col$sex)
  rownames(sex) <- df_col$sex
  row_ha_sex <- rowAnnotation(sex = sex,
                              na_col = "gray80",
                              annotation_legend_param = legends_param,
                              simple_anno_size = unit(0.2, "cm"),
                              col = list(sex = sex_colors),
                              annotation_name_gp = gpar(fontsize = 0))
  
  time <- as.matrix(df_col$time)
  rownames(time) <- df_col$time
  row_ha_time <- rowAnnotation(time = time,
                               na_col = "gray80",
                               annotation_legend_param = legends_param,
                               simple_anno_size = unit(0.2, "cm"),
                               col = list(time = Fetal_time_colors),
                               annotation_name_gp = gpar(fontsize = 0))
  
  group <- as.matrix(df_col$group)
  rownames(group) <- df_col$group
  row_ha_group <- rowAnnotation(group = group,
                                na_col = "gray80",
                                annotation_legend_param = legends_param,
                                simple_anno_size = unit(0.2, "cm"),
                                annotation_name_gp = gpar(fontsize = 0))
  
  cellType <- as.matrix(df_col$cellType)
  rownames(cellType) <- df_col$cellType
  row_ha_cellType <- rowAnnotation(cellType = cellType,
                                   na_col = "gray80",
                                   annotation_legend_param = legends_param,
                                   simple_anno_size = unit(0.2, "cm"),
                                   col = list(sex = Fetal_cellType_colors),
                                   annotation_name_gp = gpar(fontsize = 0))
  
   if (is.null(title)) title <- paste0('Promoter methylation of ', 
                                       gene, "\nTSS: ", TSS, " (hg37)")
  ht_opt$ROW_ANNO_PADDING = unit(c(0.2, 0.2), "points")
  h <-  Heatmap(mat,
                column_title = title,
                name = 'Met',
                row_split = df_col$cellType,
                cluster_row_slices = FALSE,
                row_gap = unit(c(2), "mm"),
                row_title_gp = gpar(fontsize = 5, fontface = "bold"),
                column_title_gp = gpar(fontsize = 6),
                row_title_rot = 90,
                col = colorRamp2(c(1:100),colorRampPalette(
                  c("moccasin","dodgerblue4"))(100)),
                na_col = "gray80",
                clustering_method_rows = "ward.D",
                cluster_rows = FALSE,
                cluster_columns = FALSE,
                show_row_names = TRUE,
                show_column_names = TRUE,
                show_heatmap_legend = TRUE,
                show_row_dend = FALSE,
                row_names_gp = gpar(fontsize = fs),
                column_names_gp = gpar(fontsize = col_fs),
                column_names_side = "bottom",
                heatmap_legend_param = legends_param,
                width = unit(h_width, "cm"),
                height = unit(h_height, "cm"),
                row_names_side = "left")
  
  print(h) 
}
```


### PAGE1

```{r}
gene <- "PAGE1"

A <- cowplot::plot_grid(grid.grabExpr(
  promoter_methylation_by_gene(gene = gene,
                                   sex = "F",
                                   h_width = 2,
                                   h_height = 5,
                                   fs  = 0,
                                   col_fs = 2,
                                   title = paste0("Female fetal germ cells\n(", gene, ')'))))

B <- cowplot::plot_grid(grid.grabExpr(
  promoter_methylation_by_gene(gene = gene,
                                   sex = "M",
                                   h_width = 2,
                                   h_height = 5,
                                   fs  = 0,
                                   col_fs = 2,
                                    title = paste0("Male fetal germ cells\n(", gene, ')'))))

fig <- cowplot::plot_grid(A, B, 
                   # rel_widths = c(2,1),
                   # scale = c(1, 0.8),
                   # label_size = 10, 
                   nrow = 1,
                   align = "h")
fig
```


```{r, eval = FALSE, echo=FALSE}
pdf(file = "figs/PAGE1_fetal_meth.pdf", height = 2.5, width = 5)
fig
dev.off()
```

### GAGE2A

```{r}
gene <- "GAGE2A"

A <- cowplot::plot_grid(grid.grabExpr(
  promoter_methylation_by_gene(gene = gene,
                                   sex = "F",
                                   h_width = 2,
                                   h_height = 5,
                                   nt_up = 100,
                                   fs  = 0,
                                   col_fs = 2,
                                   title = paste0("Female fetal germ cells\n(", gene, ')'))))

B <- cowplot::plot_grid(grid.grabExpr(
  promoter_methylation_by_gene(gene = gene,
                                   sex = "M",
                                   h_width = 2,
                                   h_height = 5,
                               nt_up = 100,
                                   fs  = 0,
                                   col_fs = 2,
                                    title = paste0("Male fetal germ cells\n(", gene, ')'))))

fig <- cowplot::plot_grid(A, B, 
                   # rel_widths = c(2,1),
                   # scale = c(1, 0.8),
                   # label_size = 10, 
                   nrow = 1,
                   align = "h")

fig
```
```{r, eval = FALSE, echo=FALSE}
pdf(file = "figs/GAGE2A_fetal_meth.pdf", height = 2.5, width = 5)
fig
dev.off()
```



# Gene function

```{r, warning=FALSE}
gmt_file <- msigdbr(species = "Homo sapiens", category = "C5", 
                    subcategory = "BP") %>%
  dplyr::select(gs_name, entrez_gene) %>%
  dplyr::rename(ont = gs_name, gene = entrez_gene) 

padj_thr <- 0.05
maxGSSize <- 200

all_genes <- all_genes %>% 
  mutate(ENTREZID = mapIds(org.Hs.eg.db, external_gene_name , "ENTREZID", 
                           "SYMBOL") %>% unname()) %>% 
  filter(!is.na(ENTREZID))

CT_met_X <- all_genes %>% 
  filter(CT_gene_type == "CT_gene") %>% 
  filter(regulated_by_methylation, X_linked)

egmt <- enricher(gene          = CT_met_X$ENTREZID,
                 universe      = all_genes$ENTREZID,
                 pAdjustMethod = "BH",
                 minGSSize = 10,
                 maxGSSize = maxGSSize,
                 pvalueCutoff  = 1,
                 qvalueCutoff = 1,
                 TERM2GENE = gmt_file)

egmt_tbl <- as_tibble(egmt) %>% filter(p.adjust < 0.05)

IDs <- egmt_tbl[, which(colnames(egmt_tbl) == "geneID")]
tested_ID <- strsplit(IDs$geneID, "/")
egmt_tbl$gene_names <- sapply(tested_ID, function(ids){
  paste(mapIds(org.Hs.eg.db, ids, "SYMBOL", "ENTREZID"), collapse = '/')})


egmt_tbl_CT_met_X <- egmt_tbl

CT_not_Xmet <- all_genes %>% 
  filter(CT_gene_type == "CT_gene") %>% 
  filter(!external_gene_name %in% CT_met_X$external_gene_name)


egmt <- enricher(gene          = CT_not_Xmet$ENTREZID,
                 universe      = all_genes$ENTREZID,
                 pAdjustMethod = "BH",
                 minGSSize = 10,
                 maxGSSize = maxGSSize,
                 pvalueCutoff  = 1,
                 qvalueCutoff = 1,
                 TERM2GENE     = gmt_file)

egmt_tbl <- as_tibble(egmt) %>% filter(p.adjust < 0.05)

IDs <- egmt_tbl[, which(colnames(egmt_tbl) == "geneID")]
tested_ID <- strsplit(IDs$geneID, "/")
egmt_tbl$gene_names <- sapply(tested_ID, function(ids){
  paste(mapIds(org.Hs.eg.db, ids, "SYMBOL", "ENTREZID"), collapse = '/')})

egmt_tbl_CT_not_Xmet <- egmt_tbl

GO_2_groups <- rbind(as_tibble(egmt_tbl_CT_met_X) %>% 
                       filter(p.adjust < padj_thr) %>% 
                       dplyr::select(Description, Count, p.adjust) %>% 
                       mutate(Group = "Methylation\nX-linked"),
                     as_tibble(egmt_tbl_CT_not_Xmet) %>% 
                       filter(p.adjust < padj_thr) %>% 
                       dplyr::select(Description, Count, p.adjust) %>% 
                       mutate(Group = "Other\nCT genes")) %>% 
  dplyr::select(-Count) %>% 
  pivot_wider(names_from = Group, values_from = p.adjust) %>% 
  pivot_longer(names_to = "Group", values_to = "padjusted", -Description) %>% 
  left_join(rbind(as_tibble(egmt_tbl_CT_met_X) %>% 
                    filter(p.adjust < padj_thr) %>% 
                    dplyr::select(Description, Count) %>% 
                    mutate(Group = "Methylation\nX-linked"),
                  as_tibble(egmt_tbl_CT_not_Xmet) %>% 
                    filter(p.adjust < padj_thr) %>% 
                    dplyr::select(Description, Count) %>% 
                    mutate(Group = "Other\nCT genes"))) %>% 
  mutate(group = factor(Group, levels = c("Methylation\nX-linked", 
                                          "Other\nCT genes"))) %>% 
  mutate(Description = gsub("GOBP_", x = Description, replace = ''))

sorted_GO <- GO_2_groups %>% 
  arrange(group) %>%  
  pull(Description) %>% unique() 

GO_2_groups <- GO_2_groups %>% 
  arrange(group) %>% 
  filter(Count >= 1) 

GO_2_groups$Description <- 
  factor(GO_2_groups$Description, levels = rev(unique(sorted_GO)))

# redundant GO terms (or unrelevant)
redundant <- c("NCRNA_3_END_PROCESSING",
               "RNA_3_END_PROCESSING",
               "SNRNA_METABOLIC_PROCESS",
               "MALE_MEIOTIC_NUCLEAR_DIVISION",
               "DNA_METHYLATION",
               "DNA_METHYLATION_OR_DEMETHYLATION",
               "REGULATION_OF_HISTONE_H3_K4_METHYLATION",
               "REGULATION_OF_BIOLOGICAL_PROCESS_INVOLVED_IN_SYMBIOTIC_INTERACTION",
               "NEGATIVE_REGULATION_OF_MRNA_CATABOLIC_PROCESS",
               "NEGATIVE_REGULATION_OF_RNA_CATABOLIC_PROCESS",
               "REGULATION_OF_HISTONE_METHYLATION",
               "NEGATIVE_REGULATION_OF_MRNA_METABOLIC_PROCESS",
               "HOMOLOGOUS_RECOMBINATION",
               "CHROMOSOME_ORGANIZATION_INVOLVED_IN_MEIOTIC_CELL_CYCLE",
               "HOMOLOGOUS_CHROMOSOME_SEGREGATION",
               "POSITIVE_REGULATION_OF_MEIOTIC_NUCLEAR_DIVISION",
               "REGULATION_OF_MEIOTIC_NUCLEAR_DIVISION",
               "POSITIVE_REGULATION_OF_NUCLEAR_DIVISION",
               "CELL_CELL_RECOGNITION", "POSITIVE_REGULATION_OF_REPRODUCTIVE_PROCESS",
               "REGULATION_OF_MEIOTIC_CELL_CYCLE",
               "REGULATION_OF_VIRAL_ENTRY_INTO_HOST_CELL",
               "MALE_MEIOSIS_I",
               "POSITIVE_REGULATION_OF_MEIOTIC_CELL_CYCLE",
               "HISTONE_H3_K4_TRIMETHYLATION")


GO_2_groups_not_redundant <- GO_2_groups %>% 
  arrange(group) %>% 
  filter(Count > 1) %>% 
  filter(!Description %in% redundant)

  
GO_2_groups_not_redundant$Description <- 
  factor(GO_2_groups_not_redundant$Description, 
         levels = rev(unique(GO_2_groups_not_redundant$Description)))
  


GO_terms_fig <- GO_2_groups_not_redundant %>% 
  arrange(Description) %>% 
  ggplot(aes(x = group, y = Description, color = padjusted, size = Count)) +
  geom_point() +
  scale_color_continuous(low = "red", high = "blue", 
                         guide = guide_colorbar(reverse = TRUE))+
  #  na.value = "white") +
  scale_size_continuous(limits = c(1, max(GO_2_groups_not_redundant$Count, na.rm = TRUE)), range = c(0, 3))+
  theme(axis.text.y = element_text(size = 5),
        axis.text.x = element_text(size = 6, face = "bold", vjust = 0.5),
        axis.ticks = element_line(linewidth = 0.1),
        legend.key.size = unit(4, "mm"),
        aspect.ratio = 1,
        legend.text = element_text(size = 7),
        legend.title = element_text(size = 7)) +
  xlab('') +
  ylab('')
GO_terms_fig
```

```{r, eval = FALSE, echo=FALSE}
pdf(file = paste0("figs/GO_terms_2_groups_rm_redundancy.pdf"), height = 3)
GO_terms_fig
dev.off()
```


# SessionInfo

```{r}
sessionInfo()
```

